// @ariakit/react-core/cjs/composite/composite-overflow.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _XPWK4QRCcjs = require('../__chunks/XPWK4QRC.cjs');
require('../__chunks/6SC6IEJJ.cjs');
require('../__chunks/6NFJUQ7Q.cjs');
require('../__chunks/UI5KWBYV.cjs');
require('../__chunks/L634CRNJ.cjs');
require('../__chunks/3PV6Y4XH.cjs');
require('../__chunks/WWNMQZJA.cjs');
require('../__chunks/CXFFS6H4.cjs');
require('../__chunks/W32FX7DM.cjs');
require('../__chunks/ERSQB5XO.cjs');
require('../__chunks/5RIUFRY2.cjs');
require('../__chunks/HMJY2Y4H.cjs');
require('../__chunks/JQCSYNKY.cjs');
require('../__chunks/MXT5U5PP.cjs');
require('../__chunks/JLMLKTUB.cjs');
require('../__chunks/LCWSLOYU.cjs');
require('../__chunks/V7JASX7S.cjs');
require('../__chunks/56NPFF7P.cjs');
require('../__chunks/GASMO7Y4.cjs');
require('../__chunks/O5T5YIJ4.cjs');
require('../__chunks/26P4PLHI.cjs');
require('../__chunks/YVOQ2BCB.cjs');
require('../__chunks/4SKDRUBR.cjs');
require('../__chunks/AQ2LO7WV.cjs');
require('../__chunks/DRXSFB5O.cjs');
require('../__chunks/EM3Q5QVV.cjs');
require('../__chunks/XRHNSFIK.cjs');
require('../__chunks/4IRVZ6J3.cjs');
require('../__chunks/7X7JMIHX.cjs');
require('../__chunks/WDV6EJ2U.cjs');
require('../__chunks/VFNUZFXS.cjs');
require('../__chunks/BB34VMZX.cjs');
require('../__chunks/T3QB4FR3.cjs');


var _RDNUVX4Vcjs = require('../__chunks/RDNUVX4V.cjs');




var _SQLDFLTEcjs = require('../__chunks/SQLDFLTE.cjs');


var _4KGS3DOWcjs = require('../__chunks/4KGS3DOW.cjs');
require('../__chunks/TWX5YNTH.cjs');

// src/composite/composite-overflow.tsx
var TagName = "div";
var hiddenStyle = {
  opacity: 0,
  pointerEvents: "none"
};
var useCompositeOverflow = _SQLDFLTEcjs.createHook.call(void 0, function useCompositeOverflow2({
  store,
  wrapperProps: wrapperPropsProp,
  portal = false,
  ...props
}) {
  const onFocusProp = props.onFocus;
  const onFocus = _4KGS3DOWcjs.useEvent.call(void 0, (event) => {
    onFocusProp == null ? void 0 : onFocusProp(event);
    if (event.defaultPrevented) return;
    store.show();
  });
  const mounted = _RDNUVX4Vcjs.useStoreState.call(void 0, store, "mounted");
  const getStyle = (styleProp) => mounted ? styleProp : { ...hiddenStyle, ...styleProp };
  const wrapperProps = {
    ...wrapperPropsProp,
    style: getStyle(wrapperPropsProp == null ? void 0 : wrapperPropsProp.style)
  };
  props = {
    role: "presentation",
    ...props,
    onFocus
  };
  props = _XPWK4QRCcjs.usePopover.call(void 0, {
    store,
    focusable: false,
    alwaysVisible: true,
    wrapperProps,
    portal,
    ...props
  });
  return props;
});
var CompositeOverflow = _SQLDFLTEcjs.forwardRef.call(void 0, function CompositeOverflow2(props) {
  const htmlProps = useCompositeOverflow(props);
  return _SQLDFLTEcjs.createElement.call(void 0, TagName, htmlProps);
});



exports.CompositeOverflow = CompositeOverflow; exports.useCompositeOverflow = useCompositeOverflow;
