// @ariakit/core/cjs/utils/misc.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});






















var _4TRESXTOcjs = require('../__chunks/4TRESXTO.cjs');























exports.afterPaint = _4TRESXTOcjs.afterPaint; exports.applyState = _4TRESXTOcjs.applyState; exports.beforePaint = _4TRESXTOcjs.beforePaint; exports.chain = _4TRESXTOcjs.chain; exports.cx = _4TRESXTOcjs.cx; exports.defaultValue = _4TRESXTOcjs.defaultValue; exports.disabledFromElement = _4TRESXTOcjs.disabledFromElement; exports.disabledFromProps = _4TRESXTOcjs.disabledFromProps; exports.getKeys = _4TRESXTOcjs.getKeys; exports.hasOwnProperty = _4TRESXTOcjs.hasOwnProperty; exports.identity = _4TRESXTOcjs.identity; exports.invariant = _4TRESXTOcjs.invariant; exports.isEmpty = _4TRESXTOcjs.isEmpty; exports.isFalsyBooleanCallback = _4TRESXTOcjs.isFalsyBooleanCallback; exports.isInteger = _4TRESXTOcjs.isInteger; exports.isObject = _4TRESXTOcjs.isObject; exports.noop = _4TRESXTOcjs.noop; exports.normalizeString = _4TRESXTOcjs.normalizeString; exports.omit = _4TRESXTOcjs.omit; exports.pick = _4TRESXTOcjs.pick; exports.removeUndefinedValues = _4TRESXTOcjs.removeUndefinedValues; exports.shallowEqual = _4TRESXTOcjs.shallowEqual;
