// @ariakit/react-core/cjs/tag/tag-input.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _5ABIDBW2cjs = require('../__chunks/5ABIDBW2.cjs');


var _DX5OUXRCcjs = require('../__chunks/DX5OUXRC.cjs');
require('../__chunks/2XVCJUVB.cjs');
require('../__chunks/QM74NV6D.cjs');
require('../__chunks/EOGOO37X.cjs');
require('../__chunks/HAR6UDZU.cjs');
require('../__chunks/BB34VMZX.cjs');
require('../__chunks/T3QB4FR3.cjs');


var _RDNUVX4Vcjs = require('../__chunks/RDNUVX4V.cjs');
require('../__chunks/ZYZTL3IQ.cjs');




var _SQLDFLTEcjs = require('../__chunks/SQLDFLTE.cjs');




var _4KGS3DOWcjs = require('../__chunks/4KGS3DOW.cjs');
require('../__chunks/TWX5YNTH.cjs');

// src/tag/tag-input.tsx
var _array = require('@ariakit/core/utils/array');



var _dom = require('@ariakit/core/utils/dom');
var _events = require('@ariakit/core/utils/events');
var _misc = require('@ariakit/core/utils/misc');
var _undo = require('@ariakit/core/utils/undo');
var TagName = "input";
var DEFAULT_DELIMITER = ["\n", ";", ",", /\s/];
function getDelimiters(delimiter, defaultDelimiter = DEFAULT_DELIMITER) {
  const finalDelimiter = delimiter === void 0 ? defaultDelimiter : delimiter;
  if (!finalDelimiter) return [];
  return _array.toArray.call(void 0, finalDelimiter);
}
function splitValueByDelimiter(value, delimiters) {
  for (const delimiter of delimiters) {
    let match = value.match(delimiter);
    while ((match == null ? void 0 : match.index) === 0) {
      value = value.slice(match[0].length);
      match = value.match(delimiter);
    }
    if (!match) continue;
    return value.split(delimiter);
  }
  return [];
}
var useTagInput = _SQLDFLTEcjs.createHook.call(void 0, 
  function useTagInput2({
    store,
    tabbable = true,
    delimiter,
    addValueOnPaste = true,
    addValueOnChange = true,
    setValueOnChange = true,
    removeOnBackspace = true,
    ...props
  }) {
    const context = _5ABIDBW2cjs.useTagContext.call(void 0, );
    store = store || context;
    _misc.invariant.call(void 0, 
      store,
      process.env.NODE_ENV !== "production" && "TagInput must receive a `store` prop or be wrapped in a TagProvider component."
    );
    const value = _RDNUVX4Vcjs.useStoreState.call(void 0, store, "value");
    const onPasteProp = props.onPaste;
    const addValueOnPasteProp = _4KGS3DOWcjs.useBooleanEvent.call(void 0, addValueOnPaste);
    const onPaste = _4KGS3DOWcjs.useEvent.call(void 0, (event) => {
      onPasteProp == null ? void 0 : onPasteProp(event);
      if (event.defaultPrevented) return;
      const text = event.clipboardData.getData("text");
      const delimiters = getDelimiters(delimiter);
      const values = splitValueByDelimiter(text.trim(), delimiters).map((value2) => value2.trim()).filter((value2) => value2 !== "");
      const eventWithValues = Object.assign(event, { values });
      if (!addValueOnPasteProp(eventWithValues)) return;
      if (!values.length) return;
      event.preventDefault();
      for (const tagValue of values) {
        store.addValue(tagValue);
      }
    });
    const onChangeProp = props.onChange;
    const setValueOnChangeProp = _4KGS3DOWcjs.useBooleanEvent.call(void 0, setValueOnChange);
    const addValueOnChangeProp = _4KGS3DOWcjs.useBooleanEvent.call(void 0, addValueOnChange);
    const onChange = _4KGS3DOWcjs.useEvent.call(void 0, (event) => {
      onChangeProp == null ? void 0 : onChangeProp(event);
      if (event.defaultPrevented) return;
      if (!store) return;
      const { value: prevValue } = store.getState();
      const inputType = _events.getInputType.call(void 0, event);
      const currentTarget = event.currentTarget;
      const { start, end } = _dom.getTextboxSelection.call(void 0, currentTarget);
      const { value: value2 } = currentTarget;
      if (setValueOnChangeProp(event)) {
        void _undo.UndoManager.execute(() => {
          store.setValue(value2);
          queueMicrotask(() => {
            _dom.setSelectionRange.call(void 0, currentTarget, start, end);
          });
          if (value2 === prevValue) return;
          return () => store.setValue(prevValue);
        }, inputType);
      }
      const isTrailingCaret = start === end && start === value2.length;
      if (isTrailingCaret) {
        const delimiters = getDelimiters(delimiter);
        let values = splitValueByDelimiter(value2, delimiters);
        const trailingvalue = values.pop() || "";
        values = values.map((value3) => value3.trim()).filter((value3) => value3 !== "");
        const eventWithValues = Object.assign(event, { values });
        if (values.length && addValueOnChangeProp(eventWithValues)) {
          event.preventDefault();
          for (const tagValue of values) {
            store.addValue(tagValue);
          }
          void _undo.UndoManager.execute(() => {
            store.setValue(trailingvalue);
            if (trailingvalue === prevValue) return;
            return () => store.setValue(prevValue);
          }, inputType);
        }
      }
    });
    const onKeyDownProp = props.onKeyDown;
    const removeOnBackspaceProp = _4KGS3DOWcjs.useBooleanEvent.call(void 0, removeOnBackspace);
    const onKeyDown = _4KGS3DOWcjs.useEvent.call(void 0, (event) => {
      onKeyDownProp == null ? void 0 : onKeyDownProp(event);
      if (event.defaultPrevented) return;
      if (event.key === "Backspace" && removeOnBackspaceProp(event)) {
        const { start, end } = _dom.getTextboxSelection.call(void 0, event.currentTarget);
        const isLeadingCaret = start === end && start === 0;
        if (!isLeadingCaret) return;
        store.setValues((values) => {
          if (!values.length) return values;
          return values.slice(0, -1);
        });
      }
    });
    props = {
      value,
      ...props,
      ref: _4KGS3DOWcjs.useMergeRefs.call(void 0, store.setInputElement, props.ref),
      onPaste,
      onChange,
      onKeyDown
    };
    props = _DX5OUXRCcjs.useCompositeItem.call(void 0, { store, tabbable, ...props });
    return props;
  }
);
var TagInput = _SQLDFLTEcjs.forwardRef.call(void 0, function TagInput2(props) {
  const htmlProps = useTagInput(props);
  return _SQLDFLTEcjs.createElement.call(void 0, TagName, htmlProps);
});



exports.TagInput = TagInput; exports.useTagInput = useTagInput;
