// @ariakit/core/cjs/tag/tag-store.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _CTEQSOKZcjs = require('../__chunks/CTEQSOKZ.cjs');


var _DKSHIUOKcjs = require('../__chunks/DKSHIUOK.cjs');
require('../__chunks/GDZQUFNP.cjs');
require('../__chunks/VC22SYOD.cjs');
require('../__chunks/I4CBIWTX.cjs');




var _WMTGHDLEcjs = require('../__chunks/WMTGHDLE.cjs');



var _4TRESXTOcjs = require('../__chunks/4TRESXTO.cjs');

// src/tag/tag-store.ts
function createTagStore(props = {}) {
  var _a;
  const syncState = (_a = props.store) == null ? void 0 : _a.getState();
  const composite = _DKSHIUOKcjs.createCompositeStore.call(void 0, props);
  const initialState = {
    ...composite.getState(),
    inputElement: _4TRESXTOcjs.defaultValue.call(void 0, syncState == null ? void 0 : syncState.inputElement, null),
    labelElement: _4TRESXTOcjs.defaultValue.call(void 0, syncState == null ? void 0 : syncState.labelElement, null),
    value: _4TRESXTOcjs.defaultValue.call(void 0, props.value, syncState == null ? void 0 : syncState.value, props.defaultValue, ""),
    values: _4TRESXTOcjs.defaultValue.call(void 0, 
      props.values,
      syncState == null ? void 0 : syncState.values,
      props.defaultValues,
      []
    )
  };
  const tag = _WMTGHDLEcjs.createStore.call(void 0, initialState, composite, props.store);
  _WMTGHDLEcjs.setup.call(void 0, 
    tag,
    () => _WMTGHDLEcjs.sync.call(void 0, tag, ["inputElement", "activeId"], (state) => {
      if (!state.inputElement) return;
      if (state.activeId !== void 0) return;
      tag.setState("activeId", state.inputElement.id);
    })
  );
  const setValues = (values) => {
    const { values: previousValues } = tag.getState();
    void _CTEQSOKZcjs.UndoManager.execute(() => {
      let changed = true;
      tag.setState("values", (prev) => {
        const next = _4TRESXTOcjs.applyState.call(void 0, values, prev);
        if (next === prev) {
          changed = false;
        }
        return next;
      });
      if (!changed) return;
      return () => {
        var _a2;
        tag.setState("values", previousValues);
        composite.move((_a2 = tag.getState().inputElement) == null ? void 0 : _a2.id);
      };
    });
  };
  return {
    ...composite,
    ...tag,
    setInputElement: (inputElement) => tag.setState("inputElement", inputElement),
    setLabelElement: (labelElement) => tag.setState("labelElement", labelElement),
    setValue: (value) => tag.setState("value", value),
    resetValue: () => tag.setState("value", initialState.value),
    setValues,
    addValue: (value) => {
      setValues((values) => {
        if (values.includes(value)) return values;
        return [...values, value];
      });
    },
    removeValue: (value) => setValues((values) => values.filter((v) => v !== value))
  };
}


exports.createTagStore = createTagStore;
