// @ariakit/react-core/cjs/form/form-checkbox.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _7Y3JA54Qcjs = require('../__chunks/7Y3JA54Q.cjs');


var _5CBRED7Acjs = require('../__chunks/5CBRED7A.cjs');
require('../__chunks/QM74NV6D.cjs');


var _SQEFFBAScjs = require('../__chunks/SQEFFBAS.cjs');


var _RPW5L3AEcjs = require('../__chunks/RPW5L3AE.cjs');
require('../__chunks/HAR6UDZU.cjs');
require('../__chunks/BB34VMZX.cjs');
require('../__chunks/T3QB4FR3.cjs');
require('../__chunks/IRJEBRAW.cjs');
require('../__chunks/QB525TJG.cjs');
require('../__chunks/RDNUVX4V.cjs');
require('../__chunks/ZYZTL3IQ.cjs');





var _SQLDFLTEcjs = require('../__chunks/SQLDFLTE.cjs');
require('../__chunks/4KGS3DOW.cjs');
require('../__chunks/TWX5YNTH.cjs');

// src/form/form-checkbox.tsx
var _misc = require('@ariakit/core/utils/misc');
var TagName = "input";
var useFormCheckbox = _SQLDFLTEcjs.createHook.call(void 0, 
  function useFormCheckbox2({
    store,
    name: nameProp,
    value,
    checked,
    defaultChecked,
    ...props
  }) {
    const context = _5CBRED7Acjs.useFormContext.call(void 0, );
    store = store || context;
    _misc.invariant.call(void 0, 
      store,
      process.env.NODE_ENV !== "production" && "FormCheckbox must be wrapped in a Form component."
    );
    const name = String(nameProp);
    const checkboxStore = _SQEFFBAScjs.useCheckboxStore.call(void 0, {
      value: store.useValue(name),
      setValue: (value2) => store == null ? void 0 : store.setValue(name, value2)
    });
    props = _RPW5L3AEcjs.useCheckbox.call(void 0, { store: checkboxStore, value, checked, ...props });
    props = _7Y3JA54Qcjs.useFormControl.call(void 0, {
      store,
      name,
      "aria-labelledby": void 0,
      ...props
    });
    return props;
  }
);
var FormCheckbox = _SQLDFLTEcjs.memo.call(void 0, 
  _SQLDFLTEcjs.forwardRef.call(void 0, function FormCheckbox2(props) {
    const htmlProps = useFormCheckbox(props);
    return _SQLDFLTEcjs.createElement.call(void 0, TagName, htmlProps);
  })
);



exports.FormCheckbox = FormCheckbox; exports.useFormCheckbox = useFormCheckbox;
