// @ariakit/react-core/cjs/form/form-description.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _5CBRED7Acjs = require('../__chunks/5CBRED7A.cjs');


var _QM74NV6Dcjs = require('../__chunks/QM74NV6D.cjs');
require('../__chunks/ZYZTL3IQ.cjs');





var _SQLDFLTEcjs = require('../__chunks/SQLDFLTE.cjs');



var _4KGS3DOWcjs = require('../__chunks/4KGS3DOW.cjs');
require('../__chunks/TWX5YNTH.cjs');

// src/form/form-description.tsx
var _misc = require('@ariakit/core/utils/misc');
var _react = require('react');
var TagName = "div";
var useFormDescription = _SQLDFLTEcjs.createHook.call(void 0, 
  function useFormDescription2({
    store,
    name: nameProp,
    getItem: getItemProp,
    ...props
  }) {
    const context = _5CBRED7Acjs.useFormContext.call(void 0, );
    store = store || context;
    _misc.invariant.call(void 0, 
      store,
      process.env.NODE_ENV !== "production" && "FormDescription must be wrapped in a Form component."
    );
    const id = _4KGS3DOWcjs.useId.call(void 0, props.id);
    const ref = _react.useRef.call(void 0, null);
    const name = String(nameProp);
    const getItem = _react.useCallback.call(void 0, 
      (item) => {
        const nextItem = {
          ...item,
          id: id || item.id,
          name,
          type: "description"
        };
        if (getItemProp) {
          return getItemProp(nextItem);
        }
        return nextItem;
      },
      [id, name, getItemProp]
    );
    props = {
      ...props,
      id,
      ref: _4KGS3DOWcjs.useMergeRefs.call(void 0, ref, props.ref)
    };
    props = _QM74NV6Dcjs.useCollectionItem.call(void 0, { store, ...props, getItem });
    return props;
  }
);
var FormDescription = _SQLDFLTEcjs.memo.call(void 0, 
  _SQLDFLTEcjs.forwardRef.call(void 0, function FormDescription2(props) {
    const htmlProps = useFormDescription(props);
    return _SQLDFLTEcjs.createElement.call(void 0, TagName, htmlProps);
  })
);



exports.FormDescription = FormDescription; exports.useFormDescription = useFormDescription;
