// @ariakit/react-core/cjs/form/form-submit.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _5CBRED7Acjs = require('../__chunks/5CBRED7A.cjs');


var _F6OAWDMCcjs = require('../__chunks/F6OAWDMC.cjs');
require('../__chunks/HAR6UDZU.cjs');
require('../__chunks/BB34VMZX.cjs');
require('../__chunks/T3QB4FR3.cjs');


var _RDNUVX4Vcjs = require('../__chunks/RDNUVX4V.cjs');
require('../__chunks/ZYZTL3IQ.cjs');




var _SQLDFLTEcjs = require('../__chunks/SQLDFLTE.cjs');
require('../__chunks/4KGS3DOW.cjs');
require('../__chunks/TWX5YNTH.cjs');

// src/form/form-submit.tsx
var _misc = require('@ariakit/core/utils/misc');
var TagName = "button";
var useFormSubmit = _SQLDFLTEcjs.createHook.call(void 0, 
  function useFormSubmit2({ store, accessibleWhenDisabled = true, ...props }) {
    const context = _5CBRED7Acjs.useFormContext.call(void 0, );
    store = store || context;
    _misc.invariant.call(void 0, 
      store,
      process.env.NODE_ENV !== "production" && "FormSubmit must be wrapped in a Form component."
    );
    props = {
      type: "submit",
      disabled: _RDNUVX4Vcjs.useStoreState.call(void 0, store, "submitting"),
      ...props
    };
    props = _F6OAWDMCcjs.useButton.call(void 0, { ...props, accessibleWhenDisabled });
    return props;
  }
);
var FormSubmit = _SQLDFLTEcjs.forwardRef.call(void 0, function FormSubmit2(props) {
  const htmlProps = useFormSubmit(props);
  return _SQLDFLTEcjs.createElement.call(void 0, TagName, htmlProps);
});



exports.FormSubmit = FormSubmit; exports.useFormSubmit = useFormSubmit;
