// @ariakit/react-core/cjs/__chunks/5ABIDBW2.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _EOGOO37Xcjs = require('./EOGOO37X.cjs');


var _SQLDFLTEcjs = require('./SQLDFLTE.cjs');

// src/tag/tag-context.tsx
var _react = require('react');
var TagValueContext = _react.createContext.call(void 0, null);
var TagRemoveIdContext = _react.createContext.call(void 0, 
  null
);
var ctx = _SQLDFLTEcjs.createStoreContext.call(void 0, 
  [_EOGOO37Xcjs.CompositeContextProvider],
  [_EOGOO37Xcjs.CompositeScopedContextProvider]
);
var useTagContext = ctx.useContext;
var useTagScopedContext = ctx.useScopedContext;
var useTagProviderContext = ctx.useProviderContext;
var TagContextProvider = ctx.ContextProvider;
var TagScopedContextProvider = ctx.ScopedContextProvider;









exports.TagValueContext = TagValueContext; exports.TagRemoveIdContext = TagRemoveIdContext; exports.useTagContext = useTagContext; exports.useTagScopedContext = useTagScopedContext; exports.useTagProviderContext = useTagProviderContext; exports.TagContextProvider = TagContextProvider; exports.TagScopedContextProvider = TagScopedContextProvider;
