// @ariakit/react-core/cjs/__chunks/CGP2XJO5.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _7X7JMIHXcjs = require('./7X7JMIHX.cjs');



var _EOGOO37Xcjs = require('./EOGOO37X.cjs');


var _SQLDFLTEcjs = require('./SQLDFLTE.cjs');

// src/select/select-context.tsx
var _react = require('react');
var ctx = _SQLDFLTEcjs.createStoreContext.call(void 0, 
  [_7X7JMIHXcjs.PopoverContextProvider, _EOGOO37Xcjs.CompositeContextProvider],
  [_7X7JMIHXcjs.PopoverScopedContextProvider, _EOGOO37Xcjs.CompositeScopedContextProvider]
);
var useSelectContext = ctx.useContext;
var useSelectScopedContext = ctx.useScopedContext;
var useSelectProviderContext = ctx.useProviderContext;
var SelectContextProvider = ctx.ContextProvider;
var SelectScopedContextProvider = ctx.ScopedContextProvider;
var SelectItemCheckedContext = _react.createContext.call(void 0, false);
var SelectHeadingContext = _react.createContext.call(void 0, null);









exports.useSelectContext = useSelectContext; exports.useSelectScopedContext = useSelectScopedContext; exports.useSelectProviderContext = useSelectProviderContext; exports.SelectContextProvider = SelectContextProvider; exports.SelectScopedContextProvider = SelectScopedContextProvider; exports.SelectItemCheckedContext = SelectItemCheckedContext; exports.SelectHeadingContext = SelectHeadingContext;
