// @ariakit/react-core/cjs/__chunks/DGPPZR7J.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _WDV6EJ2Ucjs = require('./WDV6EJ2U.cjs');


var _F6OAWDMCcjs = require('./F6OAWDMC.cjs');




var _SQLDFLTEcjs = require('./SQLDFLTE.cjs');


var _4KGS3DOWcjs = require('./4KGS3DOW.cjs');

// src/dialog/dialog-dismiss.tsx
var _react = require('react');
var _jsxruntime = require('react/jsx-runtime');
var TagName = "button";
var useDialogDismiss = _SQLDFLTEcjs.createHook.call(void 0, 
  function useDialogDismiss2({ store, ...props }) {
    const context = _WDV6EJ2Ucjs.useDialogScopedContext.call(void 0, );
    store = store || context;
    const onClickProp = props.onClick;
    const onClick = _4KGS3DOWcjs.useEvent.call(void 0, (event) => {
      onClickProp == null ? void 0 : onClickProp(event);
      if (event.defaultPrevented) return;
      store == null ? void 0 : store.hide();
    });
    const children = _react.useMemo.call(void 0, 
      () => /* @__PURE__ */ _jsxruntime.jsxs.call(void 0, 
        "svg",
        {
          "aria-label": "Dismiss popup",
          display: "block",
          fill: "none",
          stroke: "currentColor",
          strokeLinecap: "round",
          strokeLinejoin: "round",
          strokeWidth: 1.5,
          viewBox: "0 0 16 16",
          height: "1em",
          width: "1em",
          children: [
            /* @__PURE__ */ _jsxruntime.jsx.call(void 0, "line", { x1: "4", y1: "4", x2: "12", y2: "12" }),
            /* @__PURE__ */ _jsxruntime.jsx.call(void 0, "line", { x1: "4", y1: "12", x2: "12", y2: "4" })
          ]
        }
      ),
      []
    );
    props = {
      "data-dialog-dismiss": "",
      children,
      ...props,
      onClick
    };
    props = _F6OAWDMCcjs.useButton.call(void 0, props);
    return props;
  }
);
var DialogDismiss = _SQLDFLTEcjs.forwardRef.call(void 0, function DialogDismiss2(props) {
  const htmlProps = useDialogDismiss(props);
  return _SQLDFLTEcjs.createElement.call(void 0, TagName, htmlProps);
});




exports.useDialogDismiss = useDialogDismiss; exports.DialogDismiss = DialogDismiss;
