// @ariakit/react-core/cjs/__chunks/EOGOO37X.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _ZYZTL3IQcjs = require('./ZYZTL3IQ.cjs');


var _SQLDFLTEcjs = require('./SQLDFLTE.cjs');

// src/composite/composite-context.tsx
var _react = require('react');
var ctx = _SQLDFLTEcjs.createStoreContext.call(void 0, 
  [_ZYZTL3IQcjs.CollectionContextProvider],
  [_ZYZTL3IQcjs.CollectionScopedContextProvider]
);
var useCompositeContext = ctx.useContext;
var useCompositeScopedContext = ctx.useScopedContext;
var useCompositeProviderContext = ctx.useProviderContext;
var CompositeContextProvider = ctx.ContextProvider;
var CompositeScopedContextProvider = ctx.ScopedContextProvider;
var CompositeItemContext = _react.createContext.call(void 0, 
  void 0
);
var CompositeRowContext = _react.createContext.call(void 0, 
  void 0
);









exports.useCompositeContext = useCompositeContext; exports.useCompositeScopedContext = useCompositeScopedContext; exports.useCompositeProviderContext = useCompositeProviderContext; exports.CompositeContextProvider = CompositeContextProvider; exports.CompositeScopedContextProvider = CompositeScopedContextProvider; exports.CompositeItemContext = CompositeItemContext; exports.CompositeRowContext = CompositeRowContext;
