// @ariakit/react-core/cjs/__chunks/IOFHSY3L.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _7X7JMIHXcjs = require('./7X7JMIHX.cjs');



var _EOGOO37Xcjs = require('./EOGOO37X.cjs');


var _SQLDFLTEcjs = require('./SQLDFLTE.cjs');

// src/combobox/combobox-context.tsx
var _react = require('react');
var ComboboxListRoleContext = _react.createContext.call(void 0, 
  void 0
);
var ctx = _SQLDFLTEcjs.createStoreContext.call(void 0, 
  [_7X7JMIHXcjs.PopoverContextProvider, _EOGOO37Xcjs.CompositeContextProvider],
  [_7X7JMIHXcjs.PopoverScopedContextProvider, _EOGOO37Xcjs.CompositeScopedContextProvider]
);
var useComboboxContext = ctx.useContext;
var useComboboxScopedContext = ctx.useScopedContext;
var useComboboxProviderContext = ctx.useProviderContext;
var ComboboxContextProvider = ctx.ContextProvider;
var ComboboxScopedContextProvider = ctx.ScopedContextProvider;
var ComboboxItemValueContext = _react.createContext.call(void 0, 
  void 0
);
var ComboboxItemCheckedContext = _react.createContext.call(void 0, false);










exports.ComboboxListRoleContext = ComboboxListRoleContext; exports.useComboboxContext = useComboboxContext; exports.useComboboxScopedContext = useComboboxScopedContext; exports.useComboboxProviderContext = useComboboxProviderContext; exports.ComboboxContextProvider = ComboboxContextProvider; exports.ComboboxScopedContextProvider = ComboboxScopedContextProvider; exports.ComboboxItemValueContext = ComboboxItemValueContext; exports.ComboboxItemCheckedContext = ComboboxItemCheckedContext;
