// @ariakit/react-core/cjs/__chunks/R6IE2JDF.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _UMS7MLRYcjs = require('./UMS7MLRY.cjs');


var _LXWDFBGBcjs = require('./LXWDFBGB.cjs');


var _DN4TDPKMcjs = require('./DN4TDPKM.cjs');


var _RDNUVX4Vcjs = require('./RDNUVX4V.cjs');




var _SQLDFLTEcjs = require('./SQLDFLTE.cjs');


var _4KGS3DOWcjs = require('./4KGS3DOW.cjs');

// src/menubar/menubar.tsx
var _jsxruntime = require('react/jsx-runtime');
var TagName = "div";
var useMenubar = _SQLDFLTEcjs.createHook.call(void 0, 
  function useMenubar2({
    store: storeProp,
    composite = true,
    orientation: orientationProp,
    virtualFocus,
    focusLoop,
    rtl,
    ...props
  }) {
    const context = _UMS7MLRYcjs.useMenubarProviderContext.call(void 0, );
    storeProp = storeProp || context;
    const store = _LXWDFBGBcjs.useMenubarStore.call(void 0, {
      store: storeProp,
      orientation: orientationProp,
      virtualFocus,
      focusLoop,
      rtl
    });
    const orientation = _RDNUVX4Vcjs.useStoreState.call(void 0, 
      store,
      (state) => !composite || state.orientation === "both" ? void 0 : state.orientation
    );
    props = _4KGS3DOWcjs.useWrapElement.call(void 0, 
      props,
      (element) => /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _UMS7MLRYcjs.MenubarScopedContextProvider, { value: store, children: element }),
      [store]
    );
    if (composite) {
      props = {
        role: "menubar",
        "aria-orientation": orientation,
        ...props
      };
    }
    props = _DN4TDPKMcjs.useComposite.call(void 0, { store, composite, ...props });
    return props;
  }
);
var Menubar = _SQLDFLTEcjs.forwardRef.call(void 0, function Menubar2(props) {
  const htmlProps = useMenubar(props);
  return _SQLDFLTEcjs.createElement.call(void 0, TagName, htmlProps);
});




exports.useMenubar = useMenubar; exports.Menubar = Menubar;
