// @ariakit/react-core/cjs/__chunks/U5MBFBNT.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});





var _UMS7MLRYcjs = require('./UMS7MLRY.cjs');



var _YYY7KD73cjs = require('./YYY7KD73.cjs');



var _EOGOO37Xcjs = require('./EOGOO37X.cjs');


var _SQLDFLTEcjs = require('./SQLDFLTE.cjs');

// src/menu/menu-context.tsx
var _react = require('react');
var menu = _SQLDFLTEcjs.createStoreContext.call(void 0, 
  [_EOGOO37Xcjs.CompositeContextProvider, _YYY7KD73cjs.HovercardContextProvider],
  [_EOGOO37Xcjs.CompositeScopedContextProvider, _YYY7KD73cjs.HovercardScopedContextProvider]
);
var useMenuContext = menu.useContext;
var useMenuScopedContext = menu.useScopedContext;
var useMenuProviderContext = menu.useProviderContext;
var MenuContextProvider = menu.ContextProvider;
var MenuScopedContextProvider = menu.ScopedContextProvider;
var useMenuBarContext = _UMS7MLRYcjs.useMenubarContext;
var useMenuBarScopedContext = _UMS7MLRYcjs.useMenubarScopedContext;
var useMenuBarProviderContext = _UMS7MLRYcjs.useMenubarProviderContext;
var MenuBarContextProvider = _UMS7MLRYcjs.MenubarContextProvider;
var MenuBarScopedContextProvider = _UMS7MLRYcjs.MenubarScopedContextProvider;
var MenuItemCheckedContext = _react.createContext.call(void 0, 
  void 0
);













exports.useMenuContext = useMenuContext; exports.useMenuScopedContext = useMenuScopedContext; exports.useMenuProviderContext = useMenuProviderContext; exports.MenuContextProvider = MenuContextProvider; exports.MenuScopedContextProvider = MenuScopedContextProvider; exports.useMenuBarContext = useMenuBarContext; exports.useMenuBarScopedContext = useMenuBarScopedContext; exports.useMenuBarProviderContext = useMenuBarProviderContext; exports.MenuBarContextProvider = MenuBarContextProvider; exports.MenuBarScopedContextProvider = MenuBarScopedContextProvider; exports.MenuItemCheckedContext = MenuItemCheckedContext;
