// @ariakit/react-core/cjs/__chunks/UMS7MLRY.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _EOGOO37Xcjs = require('./EOGOO37X.cjs');


var _SQLDFLTEcjs = require('./SQLDFLTE.cjs');

// src/menubar/menubar-context.tsx
var _react = require('react');
var menubar = _SQLDFLTEcjs.createStoreContext.call(void 0, 
  [_EOGOO37Xcjs.CompositeContextProvider],
  [_EOGOO37Xcjs.CompositeScopedContextProvider]
);
var useMenubarContext = menubar.useContext;
var useMenubarScopedContext = menubar.useScopedContext;
var useMenubarProviderContext = menubar.useProviderContext;
var MenubarContextProvider = menubar.ContextProvider;
var MenubarScopedContextProvider = menubar.ScopedContextProvider;
var MenuItemCheckedContext = _react.createContext.call(void 0, 
  void 0
);








exports.useMenubarContext = useMenubarContext; exports.useMenubarScopedContext = useMenubarScopedContext; exports.useMenubarProviderContext = useMenubarProviderContext; exports.MenubarContextProvider = MenubarContextProvider; exports.MenubarScopedContextProvider = MenubarScopedContextProvider; exports.MenuItemCheckedContext = MenuItemCheckedContext;
