// @ariakit/react-core/cjs/__chunks/WDV6EJ2U.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _VFNUZFXScjs = require('./VFNUZFXS.cjs');


var _SQLDFLTEcjs = require('./SQLDFLTE.cjs');

// src/dialog/dialog-context.tsx
var _react = require('react');
var ctx = _SQLDFLTEcjs.createStoreContext.call(void 0, 
  [_VFNUZFXScjs.DisclosureContextProvider],
  [_VFNUZFXScjs.DisclosureScopedContextProvider]
);
var useDialogContext = ctx.useContext;
var useDialogScopedContext = ctx.useScopedContext;
var useDialogProviderContext = ctx.useProviderContext;
var DialogContextProvider = ctx.ContextProvider;
var DialogScopedContextProvider = ctx.ScopedContextProvider;
var DialogHeadingContext = _react.createContext.call(void 0, void 0);
var DialogDescriptionContext = _react.createContext.call(void 0, void 0);









exports.useDialogContext = useDialogContext; exports.useDialogScopedContext = useDialogScopedContext; exports.useDialogProviderContext = useDialogProviderContext; exports.DialogContextProvider = DialogContextProvider; exports.DialogScopedContextProvider = DialogScopedContextProvider; exports.DialogHeadingContext = DialogHeadingContext; exports.DialogDescriptionContext = DialogDescriptionContext;
