// @wordpress/ui/build/popover/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/ui/src/popover/index.ts
var popover_exports = {};
__export(popover_exports, {
  Arrow: () => import_arrow.Arrow,
  Close: () => import_close.Close,
  Description: () => import_description.Description,
  Popup: () => import_popup.Popup,
  Portal: () => import_portal.Portal,
  Root: () => import_root.Root,
  Title: () => import_title.Title,
  Trigger: () => import_trigger.Trigger
});
module.exports = __toCommonJS(popover_exports);
var import_arrow = require("./arrow.cjs");
var import_close = require("./close.cjs");
var import_description = require("./description.cjs");
var import_popup = require("./popup.cjs");
var import_portal = require("./portal.cjs");
var import_root = require("./root.cjs");
var import_title = require("./title.cjs");
var import_trigger = require("./trigger.cjs");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Arrow,
  Close,
  Description,
  Popup,
  Portal,
  Root,
  Title,
  Trigger
});
