// @wordpress/ui/build/popover/popup.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/ui/src/popover/popup.tsx
var popup_exports = {};
__export(popup_exports, {
  Popup: () => Popup
});
module.exports = __toCommonJS(popup_exports);
var import_popover = require("@base-ui/react/popover");
var import_clsx = __toESM(require("clsx"));
var import_element = require("@wordpress/element");
var import_compose = require("@wordpress/compose");
var import_theme = require("@wordpress/theme");
var import_lock_unlock = require("../lock-unlock.cjs");

// packages/ui/src/utils/css/resets.module.css
if (typeof document !== "undefined" && process.env.NODE_ENV !== "test" && !document.head.querySelector("style[data-wp-hash='e3ae230cea']")) {
  const style = document.createElement("style");
  style.setAttribute("data-wp-hash", "e3ae230cea");
  style.appendChild(document.createTextNode("@layer wp-ui-utilities, wp-ui-components, wp-ui-compositions, wp-ui-overrides;@layer wp-ui-utilities{._336cd3e4e743482f__box-sizing{box-sizing:border-box;*,:after,:before{box-sizing:inherit}}}"));
  document.head.appendChild(style);
}
var resets_default = { "box-sizing": "_336cd3e4e743482f__box-sizing" };

// packages/ui/src/popover/popup.tsx
var import_use_deprioritized_initial_focus = require("../utils/use-deprioritized-initial-focus.cjs");
var import_render_portal_with_children = require("../utils/render-portal-with-children.cjs");
var import_context = require("./context.cjs");
var import_portal = require("./portal.cjs");

// packages/ui/src/popover/style.module.css
if (typeof document !== "undefined" && process.env.NODE_ENV !== "test" && !document.head.querySelector("style[data-wp-hash='cc851ac208']")) {
  const style = document.createElement("style");
  style.setAttribute("data-wp-hash", "cc851ac208");
  style.appendChild(document.createTextNode('@layer wp-ui-utilities, wp-ui-components, wp-ui-compositions, wp-ui-overrides;@layer wp-ui-components{._10450722b9676f78__positioner{z-index:var(--wp-ui-popover-z-index,initial)}._84e8f597bcf683b8__popup{background-color:var(--wpds-color-bg-surface-neutral-strong,#fff);border:var(--wpds-border-width-xs,1px) solid var(--wpds-color-stroke-surface-neutral,#dbdbdb);border-radius:var(--wpds-border-radius-md,4px);box-shadow:var(--wpds-elevation-md,0 2px 3px 0 #0000000d,0 4px 5px 0 #0000000a,0 12px 12px 0 #00000008,0 16px 16px 0 #00000005);color:var(--wpds-color-fg-content-neutral,#1e1e1e);font-family:var(--wpds-typography-font-family-body,-apple-system,system-ui,"Segoe UI","Roboto","Oxygen-Sans","Ubuntu","Cantarell","Helvetica Neue",sans-serif);font-size:var(--wpds-typography-font-size-md,13px);line-height:var(--wpds-typography-line-height-md,24px);outline:0;padding:var(--wpds-dimension-padding-lg,16px)}.e4d544aa033f05c6__backdrop~* ._84e8f597bcf683b8__popup{border-color:#0000}@media (forced-colors:active){._84e8f597bcf683b8__popup,.e4d544aa033f05c6__backdrop~* ._84e8f597bcf683b8__popup{border-color:CanvasText}}.ca05d3eb89321fcd__arrow{display:flex;&[data-side=top]{bottom:-8px;rotate:180deg}&[data-side=bottom]{rotate:0deg;top:-8px}&[data-side=left]{inset-inline-end:-13px;rotate:90deg}&[data-side=right]{inset-inline-start:-13px;rotate:-90deg}}._12d8edd9eb946b5f__arrow-fill{fill:var(--wpds-color-bg-surface-neutral-strong,#fff)}._6ddab482bd929dad__arrow-stroke{fill:var(--wpds-color-stroke-surface-neutral,#dbdbdb)}._6d7e4729cd96960e__title{--_gcd-heading-color:var(--wpds-color-fg-content-neutral,#1e1e1e);color:var(--wpds-color-fg-content-neutral,#1e1e1e)}.e4d544aa033f05c6__backdrop{background-color:#00000026;inset:0;position:fixed;z-index:var(--wp-ui-popover-z-index,initial)}}'));
  document.head.appendChild(style);
}
var style_default = { "positioner": "_10450722b9676f78__positioner", "popup": "_84e8f597bcf683b8__popup", "backdrop": "e4d544aa033f05c6__backdrop", "arrow": "ca05d3eb89321fcd__arrow", "arrow-fill": "_12d8edd9eb946b5f__arrow-fill", "arrow-stroke": "_6ddab482bd929dad__arrow-stroke", "title": "_6d7e4729cd96960e__title" };

// packages/ui/src/popover/popup.tsx
var import_jsx_runtime = require("react/jsx-runtime");
var ThemeProvider = (0, import_lock_unlock.unlock)(import_theme.privateApis).ThemeProvider;
var CLOSE_ATTR = "data-wp-ui-popover-close";
var Popup = (0, import_element.forwardRef)(function PopoverPopup({
  align = "center",
  alignOffset,
  anchor,
  // Matches the popup's border-radius (--wpds-border-radius-md).
  arrowPadding = 8,
  backdrop = false,
  children,
  className,
  collisionAvoidance,
  collisionBoundary,
  collisionPadding,
  portal,
  finalFocus,
  initialFocus,
  side = "bottom",
  sideOffset = 8,
  sticky,
  variant = "default",
  ...props
}, ref) {
  const { resolvedInitialFocus, popupRef } = (0, import_use_deprioritized_initial_focus.useDeprioritizedInitialFocus)({
    initialFocus,
    deprioritizedAttributes: [CLOSE_ATTR]
  });
  const mergedPopupRef = (0, import_compose.useMergeRefs)([ref, popupRef]);
  const backdropElement = backdrop ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_popover.Popover.Backdrop, { className: style_default.backdrop }) : null;
  const positioner = /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_popover.Popover.Positioner,
    {
      align,
      alignOffset,
      anchor,
      arrowPadding,
      collisionAvoidance,
      collisionBoundary,
      collisionPadding,
      side,
      sideOffset,
      sticky,
      className: (0, import_clsx.default)(resets_default["box-sizing"], style_default.positioner),
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(ThemeProvider, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_popover.Popover.Popup,
        {
          ref: mergedPopupRef,
          initialFocus: resolvedInitialFocus,
          finalFocus,
          className: (0, import_clsx.default)(
            variant !== "unstyled" && style_default.popup,
            className
          ),
          ...props,
          children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_context.PopoverValidationProvider, { children })
        }
      ) })
    }
  );
  const portalChildren = /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    backdropElement,
    positioner
  ] });
  return (0, import_render_portal_with_children.renderPortalWithChildren)(portal, /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_portal.Portal, {}), portalChildren);
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Popup
});
