// @ariakit/react-core/cjs/menu/menu-item-checkbox.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _2ORAXZT6cjs = require('../__chunks/2ORAXZT6.cjs');


var _U5MBFBNTcjs = require('../__chunks/U5MBFBNT.cjs');
require('../__chunks/UMS7MLRY.cjs');
require('../__chunks/YYY7KD73.cjs');
require('../__chunks/3VALIS7Q.cjs');
require('../__chunks/DX5OUXRC.cjs');
require('../__chunks/2XVCJUVB.cjs');
require('../__chunks/QM74NV6D.cjs');
require('../__chunks/7X7JMIHX.cjs');
require('../__chunks/WDV6EJ2U.cjs');
require('../__chunks/VFNUZFXS.cjs');
require('../__chunks/EOGOO37X.cjs');


var _SQEFFBAScjs = require('../__chunks/SQEFFBAS.cjs');


var _RPW5L3AEcjs = require('../__chunks/RPW5L3AE.cjs');
require('../__chunks/HAR6UDZU.cjs');
require('../__chunks/BB34VMZX.cjs');
require('../__chunks/T3QB4FR3.cjs');
require('../__chunks/IRJEBRAW.cjs');
require('../__chunks/QB525TJG.cjs');


var _RDNUVX4Vcjs = require('../__chunks/RDNUVX4V.cjs');
require('../__chunks/ZYZTL3IQ.cjs');





var _SQLDFLTEcjs = require('../__chunks/SQLDFLTE.cjs');


var _4KGS3DOWcjs = require('../__chunks/4KGS3DOW.cjs');
require('../__chunks/TWX5YNTH.cjs');

// src/menu/menu-item-checkbox.tsx
var _misc = require('@ariakit/core/utils/misc');
var _react = require('react');
var TagName = "div";
function getPrimitiveValue(value) {
  if (Array.isArray(value)) {
    return value.toString();
  }
  return value;
}
function getValue(storeValue, value, checked) {
  if (value === void 0) {
    if (Array.isArray(storeValue)) return storeValue;
    return !!checked;
  }
  const primitiveValue = getPrimitiveValue(value);
  if (!Array.isArray(storeValue)) {
    if (checked) {
      return primitiveValue;
    }
    return storeValue === primitiveValue ? false : storeValue;
  }
  if (checked) {
    if (storeValue.includes(primitiveValue)) {
      return storeValue;
    }
    return [...storeValue, primitiveValue];
  }
  return storeValue.filter((v) => v !== primitiveValue);
}
var useMenuItemCheckbox = _SQLDFLTEcjs.createHook.call(void 0, 
  function useMenuItemCheckbox2({
    store,
    name,
    value,
    checked,
    defaultChecked: defaultCheckedProp,
    hideOnClick = false,
    ...props
  }) {
    const context = _U5MBFBNTcjs.useMenuScopedContext.call(void 0, );
    store = store || context;
    _misc.invariant.call(void 0, 
      store,
      process.env.NODE_ENV !== "production" && "MenuItemCheckbox must be wrapped in a MenuList or Menu component"
    );
    const defaultChecked = _4KGS3DOWcjs.useInitialValue.call(void 0, defaultCheckedProp);
    _react.useEffect.call(void 0, () => {
      store == null ? void 0 : store.setValue(name, (prevValue = []) => {
        if (!defaultChecked) return prevValue;
        return getValue(prevValue, value, true);
      });
    }, [store, name, value, defaultChecked]);
    _react.useEffect.call(void 0, () => {
      if (checked === void 0) return;
      store == null ? void 0 : store.setValue(name, (prevValue) => {
        return getValue(prevValue, value, checked);
      });
    }, [store, name, value, checked]);
    const checkboxStore = _SQEFFBAScjs.useCheckboxStore.call(void 0, {
      value: _RDNUVX4Vcjs.useStoreState.call(void 0, store, (state) => state.values[name]),
      setValue(internalValue) {
        store == null ? void 0 : store.setValue(name, () => {
          if (checked === void 0) return internalValue;
          const nextValue = getValue(internalValue, value, checked);
          if (!Array.isArray(nextValue)) return nextValue;
          if (!Array.isArray(internalValue)) return nextValue;
          if (_misc.shallowEqual.call(void 0, internalValue, nextValue)) return internalValue;
          return nextValue;
        });
      }
    });
    props = {
      role: "menuitemcheckbox",
      ...props
    };
    props = _RPW5L3AEcjs.useCheckbox.call(void 0, {
      store: checkboxStore,
      name,
      value,
      checked,
      ...props
    });
    props = _2ORAXZT6cjs.useMenuItem.call(void 0, { store, hideOnClick, ...props });
    return props;
  }
);
var MenuItemCheckbox = _SQLDFLTEcjs.memo.call(void 0, 
  _SQLDFLTEcjs.forwardRef.call(void 0, function MenuItemCheckbox2(props) {
    const htmlProps = useMenuItemCheckbox(props);
    return _SQLDFLTEcjs.createElement.call(void 0, TagName, htmlProps);
  })
);



exports.MenuItemCheckbox = MenuItemCheckbox; exports.useMenuItemCheckbox = useMenuItemCheckbox;
