// @ariakit/react-core/cjs/menu/menu-context.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});











var _U5MBFBNTcjs = require('../__chunks/U5MBFBNT.cjs');
require('../__chunks/UMS7MLRY.cjs');
require('../__chunks/YYY7KD73.cjs');
require('../__chunks/7X7JMIHX.cjs');
require('../__chunks/WDV6EJ2U.cjs');
require('../__chunks/VFNUZFXS.cjs');
require('../__chunks/EOGOO37X.cjs');
require('../__chunks/ZYZTL3IQ.cjs');
require('../__chunks/SQLDFLTE.cjs');
require('../__chunks/4KGS3DOW.cjs');
require('../__chunks/TWX5YNTH.cjs');












exports.MenuBarContextProvider = _U5MBFBNTcjs.MenuBarContextProvider; exports.MenuBarScopedContextProvider = _U5MBFBNTcjs.MenuBarScopedContextProvider; exports.MenuContextProvider = _U5MBFBNTcjs.MenuContextProvider; exports.MenuItemCheckedContext = _U5MBFBNTcjs.MenuItemCheckedContext; exports.MenuScopedContextProvider = _U5MBFBNTcjs.MenuScopedContextProvider; exports.useMenuBarContext = _U5MBFBNTcjs.useMenuBarContext; exports.useMenuBarProviderContext = _U5MBFBNTcjs.useMenuBarProviderContext; exports.useMenuBarScopedContext = _U5MBFBNTcjs.useMenuBarScopedContext; exports.useMenuContext = _U5MBFBNTcjs.useMenuContext; exports.useMenuProviderContext = _U5MBFBNTcjs.useMenuProviderContext; exports.useMenuScopedContext = _U5MBFBNTcjs.useMenuScopedContext;
