// @wordpress/ui/build/alert-dialog/root.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/ui/src/alert-dialog/root.tsx
var root_exports = {};
__export(root_exports, {
  Root: () => Root
});
module.exports = __toCommonJS(root_exports);
var import_alert_dialog = require("@base-ui/react/alert-dialog");
var import_a11y = require("@wordpress/a11y");
var import_element = require("@wordpress/element");
var import_context = require("./context.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
function isThenable(value) {
  return value !== null && value !== void 0 && typeof value.then === "function";
}
function Root({
  children,
  open: openProp,
  onOpenChange,
  defaultOpen,
  onConfirm
}) {
  const [internalOpen, setInternalOpen] = (0, import_element.useState)(defaultOpen ?? false);
  const [phase, setPhase] = (0, import_element.useState)("idle");
  const [showSpinner, setShowSpinner] = (0, import_element.useState)(false);
  const [errorMessage, setErrorMessage] = (0, import_element.useState)();
  const actionsRef = (0, import_element.useRef)(null);
  const onConfirmRef = (0, import_element.useRef)(onConfirm);
  onConfirmRef.current = onConfirm;
  const phaseRef = (0, import_element.useRef)(phase);
  phaseRef.current = phase;
  const confirmIdRef = (0, import_element.useRef)(0);
  const effectiveOpen = openProp ?? internalOpen;
  (0, import_element.useEffect)(() => {
    if (effectiveOpen && phase === "closing") {
      phaseRef.current = "idle";
      setPhase("idle");
      setShowSpinner(false);
    }
  }, [effectiveOpen, phase]);
  const handleOpenChange = (0, import_element.useCallback)(
    (nextOpen, eventDetails) => {
      if (!nextOpen && phaseRef.current === "pending") {
        return;
      }
      if (!nextOpen && phaseRef.current === "idle") {
        phaseRef.current = "closing";
        setPhase("closing");
      }
      setInternalOpen(nextOpen);
      onOpenChange?.(nextOpen, eventDetails);
    },
    [onOpenChange]
  );
  const confirm = (0, import_element.useCallback)(async () => {
    if (phaseRef.current !== "idle") {
      return;
    }
    phaseRef.current = "pending";
    setPhase("pending");
    setErrorMessage(void 0);
    const id = ++confirmIdRef.current;
    try {
      const rawResult = onConfirmRef.current?.();
      if (isThenable(rawResult)) {
        setShowSpinner(true);
      }
      const result = await Promise.resolve(rawResult);
      if (confirmIdRef.current !== id) {
        return;
      }
      if (result?.error) {
        phaseRef.current = "idle";
        setPhase("idle");
        setShowSpinner(false);
        setErrorMessage(result.error);
        (0, import_a11y.speak)(result.error, "assertive");
        return;
      }
      const shouldClose = result?.close !== false;
      if (shouldClose) {
        phaseRef.current = "closing";
        setPhase("closing");
        actionsRef.current?.close();
      } else {
        phaseRef.current = "idle";
        setPhase("idle");
        setShowSpinner(false);
      }
    } catch (error) {
      if (confirmIdRef.current !== id) {
        return;
      }
      phaseRef.current = "idle";
      setPhase("idle");
      setShowSpinner(false);
      console.error(error);
    }
  }, []);
  const handleOpenChangeComplete = (0, import_element.useCallback)((open) => {
    if (!open) {
      confirmIdRef.current++;
      phaseRef.current = "idle";
      setPhase("idle");
      setShowSpinner(false);
      setErrorMessage(void 0);
    }
  }, []);
  const contextValue = (0, import_element.useMemo)(
    () => ({
      phase,
      showSpinner,
      errorMessage,
      confirm
    }),
    [phase, showSpinner, errorMessage, confirm]
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_alert_dialog.AlertDialog.Root,
    {
      open: effectiveOpen,
      defaultOpen,
      onOpenChange: handleOpenChange,
      onOpenChangeComplete: handleOpenChangeComplete,
      actionsRef,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_context.AlertDialogContext.Provider, { value: contextValue, children })
    }
  );
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Root
});
