// @wordpress/ui/build-module/alert-dialog/root.mjs
// packages/ui/src/alert-dialog/root.tsx
import { AlertDialog as _AlertDialog } from "@base-ui/react/alert-dialog";
import { speak } from "@wordpress/a11y";
import {
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState
} from "@wordpress/element";
import { AlertDialogContext } from "./context.mjs";
import { jsx } from "react/jsx-runtime";
function isThenable(value) {
  return value !== null && value !== void 0 && typeof value.then === "function";
}
function Root({
  children,
  open: openProp,
  onOpenChange,
  defaultOpen,
  onConfirm
}) {
  const [internalOpen, setInternalOpen] = useState(defaultOpen ?? false);
  const [phase, setPhase] = useState("idle");
  const [showSpinner, setShowSpinner] = useState(false);
  const [errorMessage, setErrorMessage] = useState();
  const actionsRef = useRef(null);
  const onConfirmRef = useRef(onConfirm);
  onConfirmRef.current = onConfirm;
  const phaseRef = useRef(phase);
  phaseRef.current = phase;
  const confirmIdRef = useRef(0);
  const effectiveOpen = openProp ?? internalOpen;
  useEffect(() => {
    if (effectiveOpen && phase === "closing") {
      phaseRef.current = "idle";
      setPhase("idle");
      setShowSpinner(false);
    }
  }, [effectiveOpen, phase]);
  const handleOpenChange = useCallback(
    (nextOpen, eventDetails) => {
      if (!nextOpen && phaseRef.current === "pending") {
        return;
      }
      if (!nextOpen && phaseRef.current === "idle") {
        phaseRef.current = "closing";
        setPhase("closing");
      }
      setInternalOpen(nextOpen);
      onOpenChange?.(nextOpen, eventDetails);
    },
    [onOpenChange]
  );
  const confirm = useCallback(async () => {
    if (phaseRef.current !== "idle") {
      return;
    }
    phaseRef.current = "pending";
    setPhase("pending");
    setErrorMessage(void 0);
    const id = ++confirmIdRef.current;
    try {
      const rawResult = onConfirmRef.current?.();
      if (isThenable(rawResult)) {
        setShowSpinner(true);
      }
      const result = await Promise.resolve(rawResult);
      if (confirmIdRef.current !== id) {
        return;
      }
      if (result?.error) {
        phaseRef.current = "idle";
        setPhase("idle");
        setShowSpinner(false);
        setErrorMessage(result.error);
        speak(result.error, "assertive");
        return;
      }
      const shouldClose = result?.close !== false;
      if (shouldClose) {
        phaseRef.current = "closing";
        setPhase("closing");
        actionsRef.current?.close();
      } else {
        phaseRef.current = "idle";
        setPhase("idle");
        setShowSpinner(false);
      }
    } catch (error) {
      if (confirmIdRef.current !== id) {
        return;
      }
      phaseRef.current = "idle";
      setPhase("idle");
      setShowSpinner(false);
      console.error(error);
    }
  }, []);
  const handleOpenChangeComplete = useCallback((open) => {
    if (!open) {
      confirmIdRef.current++;
      phaseRef.current = "idle";
      setPhase("idle");
      setShowSpinner(false);
      setErrorMessage(void 0);
    }
  }, []);
  const contextValue = useMemo(
    () => ({
      phase,
      showSpinner,
      errorMessage,
      confirm
    }),
    [phase, showSpinner, errorMessage, confirm]
  );
  return /* @__PURE__ */ jsx(
    _AlertDialog.Root,
    {
      open: effectiveOpen,
      defaultOpen,
      onOpenChange: handleOpenChange,
      onOpenChangeComplete: handleOpenChangeComplete,
      actionsRef,
      children: /* @__PURE__ */ jsx(AlertDialogContext.Provider, { value: contextValue, children })
    }
  );
}
export {
  Root
};
