// @ariakit/react-core/cjs/combobox/combobox-item-offscreen.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _OA2C6TXAcjs = require('../__chunks/OA2C6TXA.cjs');


var _QE7OMGCVcjs = require('../__chunks/QE7OMGCV.cjs');
require('../__chunks/3VALIS7Q.cjs');
require('../__chunks/DX5OUXRC.cjs');
require('../__chunks/2XVCJUVB.cjs');
require('../__chunks/GLYXYN6C.cjs');


var _4IRVZ6J3cjs = require('../__chunks/4IRVZ6J3.cjs');
require('../__chunks/QM74NV6D.cjs');



var _IOFHSY3Lcjs = require('../__chunks/IOFHSY3L.cjs');
require('../__chunks/7X7JMIHX.cjs');
require('../__chunks/WDV6EJ2U.cjs');
require('../__chunks/VFNUZFXS.cjs');
require('../__chunks/EOGOO37X.cjs');
require('../__chunks/HAR6UDZU.cjs');
require('../__chunks/BB34VMZX.cjs');
require('../__chunks/T3QB4FR3.cjs');
require('../__chunks/RDNUVX4V.cjs');
require('../__chunks/ZYZTL3IQ.cjs');


var _SQLDFLTEcjs = require('../__chunks/SQLDFLTE.cjs');


var _4KGS3DOWcjs = require('../__chunks/4KGS3DOW.cjs');
require('../__chunks/TWX5YNTH.cjs');

// src/combobox/combobox-item-offscreen.tsx
var _react = require('react');
var _jsxruntime = require('react/jsx-runtime');
var TagName = "div";
var itemRoleByPopupRole = {
  menu: "menuitem",
  listbox: "option",
  tree: "treeitem"
};
function getItemRole(popupRole) {
  var _a;
  const key = popupRole;
  return (_a = itemRoleByPopupRole[key]) != null ? _a : "option";
}
function useComboboxItemOffscreen({ store, value, ...props }) {
  const context = _IOFHSY3Lcjs.useComboboxScopedContext.call(void 0, );
  store = store || context;
  const offscreenProps = _OA2C6TXAcjs.useCompositeItemOffscreen.call(void 0, { store, value, ...props });
  const popupRole = _react.useContext.call(void 0, _IOFHSY3Lcjs.ComboboxListRoleContext);
  if (!offscreenProps.active) {
    return {
      ...offscreenProps,
      role: getItemRole(popupRole)
    };
  }
  return offscreenProps;
}
var ComboboxItem2 = _SQLDFLTEcjs.forwardRef.call(void 0, function ComboboxItem3({
  offscreenMode,
  offscreenRoot,
  ...props
}) {
  const { active, ref, ...rest } = useComboboxItemOffscreen({
    offscreenMode,
    offscreenRoot,
    ...props
  });
  const allProps = { ...rest, ...props, ref: _4KGS3DOWcjs.useMergeRefs.call(void 0, ref, props.ref) };
  if (active) {
    return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _QE7OMGCVcjs.ComboboxItem, { ...allProps });
  }
  const {
    store,
    value,
    hideOnClick,
    setValueOnClick,
    selectValueOnClick,
    resetValueOnSelect,
    focusOnHover,
    blurOnHoverEnd,
    moveOnKeyPress,
    getItem,
    ...htmlProps
  } = allProps;
  const Component = _4IRVZ6J3cjs.Role[TagName];
  return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, Component, { ...htmlProps });
});



exports.ComboboxItem = ComboboxItem2; exports.useComboboxItemOffscreen = useComboboxItemOffscreen;
