// @ariakit/react-core/cjs/tab/tab-list.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _BESHUAW7cjs = require('../__chunks/BESHUAW7.cjs');


var _DN4TDPKMcjs = require('../__chunks/DN4TDPKM.cjs');
require('../__chunks/2XVCJUVB.cjs');
require('../__chunks/EOGOO37X.cjs');
require('../__chunks/BB34VMZX.cjs');
require('../__chunks/T3QB4FR3.cjs');


var _RDNUVX4Vcjs = require('../__chunks/RDNUVX4V.cjs');
require('../__chunks/ZYZTL3IQ.cjs');




var _SQLDFLTEcjs = require('../__chunks/SQLDFLTE.cjs');


var _4KGS3DOWcjs = require('../__chunks/4KGS3DOW.cjs');
require('../__chunks/TWX5YNTH.cjs');

// src/tab/tab-list.tsx
var _misc = require('@ariakit/core/utils/misc');
var _jsxruntime = require('react/jsx-runtime');
var TagName = "div";
var useTabList = _SQLDFLTEcjs.createHook.call(void 0, 
  function useTabList2({ store, ...props }) {
    const context = _BESHUAW7cjs.useTabProviderContext.call(void 0, );
    store = store || context;
    _misc.invariant.call(void 0, 
      store,
      process.env.NODE_ENV !== "production" && "TabList must receive a `store` prop or be wrapped in a TabProvider component."
    );
    const orientation = _RDNUVX4Vcjs.useStoreState.call(void 0, 
      store,
      (state) => state.orientation === "both" ? void 0 : state.orientation
    );
    props = _4KGS3DOWcjs.useWrapElement.call(void 0, 
      props,
      (element) => /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _BESHUAW7cjs.TabScopedContextProvider, { value: store, children: element }),
      [store]
    );
    if (store.composite) {
      props = {
        focusable: false,
        ...props
      };
    }
    props = {
      role: "tablist",
      "aria-orientation": orientation,
      ...props
    };
    props = _DN4TDPKMcjs.useComposite.call(void 0, { store, ...props });
    return props;
  }
);
var TabList = _SQLDFLTEcjs.forwardRef.call(void 0, function TabList2(props) {
  const htmlProps = useTabList(props);
  return _SQLDFLTEcjs.createElement.call(void 0, TagName, htmlProps);
});



exports.TabList = TabList; exports.useTabList = useTabList;
