// @ariakit/react-core/cjs/toolbar/toolbar.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _FOJUOLKKcjs = require('../__chunks/FOJUOLKK.cjs');



var _W6XT2DGQcjs = require('../__chunks/W6XT2DGQ.cjs');
require('../__chunks/LRPJMEC3.cjs');


var _DN4TDPKMcjs = require('../__chunks/DN4TDPKM.cjs');
require('../__chunks/2XVCJUVB.cjs');
require('../__chunks/PXZUDDVS.cjs');
require('../__chunks/EOGOO37X.cjs');
require('../__chunks/BB34VMZX.cjs');
require('../__chunks/T3QB4FR3.cjs');


var _RDNUVX4Vcjs = require('../__chunks/RDNUVX4V.cjs');
require('../__chunks/ZYZTL3IQ.cjs');




var _SQLDFLTEcjs = require('../__chunks/SQLDFLTE.cjs');


var _4KGS3DOWcjs = require('../__chunks/4KGS3DOW.cjs');
require('../__chunks/TWX5YNTH.cjs');

// src/toolbar/toolbar.tsx
var _jsxruntime = require('react/jsx-runtime');
var TagName = "div";
var useToolbar = _SQLDFLTEcjs.createHook.call(void 0, 
  function useToolbar2({
    store: storeProp,
    orientation: orientationProp,
    virtualFocus,
    focusLoop,
    rtl,
    ...props
  }) {
    const context = _W6XT2DGQcjs.useToolbarProviderContext.call(void 0, );
    storeProp = storeProp || context;
    const store = _FOJUOLKKcjs.useToolbarStore.call(void 0, {
      store: storeProp,
      orientation: orientationProp,
      virtualFocus,
      focusLoop,
      rtl
    });
    const orientation = _RDNUVX4Vcjs.useStoreState.call(void 0, 
      store,
      (state) => state.orientation === "both" ? void 0 : state.orientation
    );
    props = _4KGS3DOWcjs.useWrapElement.call(void 0, 
      props,
      (element) => /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _W6XT2DGQcjs.ToolbarScopedContextProvider, { value: store, children: element }),
      [store]
    );
    props = {
      role: "toolbar",
      "aria-orientation": orientation,
      ...props
    };
    props = _DN4TDPKMcjs.useComposite.call(void 0, { store, ...props });
    return props;
  }
);
var Toolbar = _SQLDFLTEcjs.forwardRef.call(void 0, function Toolbar2(props) {
  const htmlProps = useToolbar(props);
  return _SQLDFLTEcjs.createElement.call(void 0, TagName, htmlProps);
});



exports.Toolbar = Toolbar; exports.useToolbar = useToolbar;
