// @wordpress/commands/build/components/command-menu.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/commands/src/components/command-menu.js
var command_menu_exports = {};
__export(command_menu_exports, {
  CommandMenu: () => CommandMenu,
  isValidIcon: () => isValidIcon
});
module.exports = __toCommonJS(command_menu_exports);
var import_cmdk = require("cmdk");
var import_clsx = __toESM(require("clsx"));
var import_data = require("@wordpress/data");
var import_element = require("@wordpress/element");
var import_i18n = require("@wordpress/i18n");
var import_components = require("@wordpress/components");
var import_keyboard_shortcuts = require("@wordpress/keyboard-shortcuts");
var import_icons = require("@wordpress/icons");
var import_store = require("../store/index.cjs");
var import_lock_unlock = require("../lock-unlock.cjs");
var import_use_recent_commands = require("./use-recent-commands.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
var { withIgnoreIMEEvents } = (0, import_lock_unlock.unlock)(import_components.privateApis);
var ITEM_ID_PREFIX = "command-palette-item-";
var inputLabel = (0, import_i18n.__)("Search commands and settings");
var CATEGORY_ICONS = {
  view: import_icons.arrowRight
};
var CATEGORY_LABELS = {
  command: (0, import_i18n.__)("Command"),
  view: (0, import_i18n.__)("View"),
  edit: (0, import_i18n.__)("Edit"),
  action: (0, import_i18n.__)("Action"),
  workflow: (0, import_i18n.__)("Workflow")
};
function isValidIcon(icon) {
  return !!icon && (typeof icon === "string" || (0, import_element.isValidElement)(icon) || typeof icon === "function" || icon instanceof import_element.Component);
}
function CommandItem({ command, search, category, valuePrefix }) {
  const { close } = (0, import_data.useDispatch)(import_store.store);
  const commandCategory = category ?? command.category;
  const label = command.searchLabel ?? command.label;
  const value = valuePrefix ? `${valuePrefix}${command.name}` : label;
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_cmdk.Command.Item,
    {
      id: `${ITEM_ID_PREFIX}${value.toLowerCase()}`,
      value,
      keywords: valuePrefix ? [...command.keywords ?? [], label] : command.keywords,
      onSelect: () => {
        (0, import_use_recent_commands.recordUsage)(command.name);
        command.callback({ close });
      },
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
        import_components.__experimentalHStack,
        {
          alignment: "left",
          className: (0, import_clsx.default)("commands-command-menu__item", {
            "has-icon": CATEGORY_ICONS[commandCategory] || command.icon
          }),
          children: [
            CATEGORY_ICONS[commandCategory] ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_icons.Icon, { icon: CATEGORY_ICONS[commandCategory] }) : isValidIcon(command.icon) && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_icons.Icon, { icon: command.icon }),
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { className: "commands-command-menu__item-label", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
              import_components.TextHighlight,
              {
                text: command.label,
                highlight: search
              }
            ) }),
            CATEGORY_LABELS[commandCategory] && /* @__PURE__ */ (0, import_jsx_runtime.jsx)("span", { className: "commands-command-menu__item-category", children: CATEGORY_LABELS[commandCategory] })
          ]
        }
      )
    },
    command.name
  );
}
function CommandMenuLoader({ name, search, hook, category, valuePrefix }) {
  const { setLoaderLoading } = (0, import_lock_unlock.unlock)((0, import_data.useDispatch)(import_store.store));
  const { isLoading: loading, commands = [] } = hook({ search }) ?? {};
  (0, import_element.useEffect)(() => {
    setLoaderLoading(name, loading);
  }, [setLoaderLoading, name, loading]);
  if (!commands.length) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children: commands.map((command) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    CommandItem,
    {
      command,
      search,
      category: command.category ?? category,
      valuePrefix
    },
    command.name
  )) });
}
function CommandMenuLoaderWrapper({ hook, ...props }) {
  const currentLoaderRef = (0, import_element.useRef)(hook);
  const [key, setKey] = (0, import_element.useState)(0);
  (0, import_element.useEffect)(() => {
    if (currentLoaderRef.current !== hook) {
      currentLoaderRef.current = hook;
      setKey((prevKey) => prevKey + 1);
    }
  }, [hook]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    CommandMenuLoader,
    {
      hook: currentLoaderRef.current,
      ...props
    },
    key
  );
}
function CommandList({ search, commands, loaders, valuePrefix }) {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    commands.map((command) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      CommandItem,
      {
        command,
        search,
        valuePrefix
      },
      command.name
    )),
    loaders.map((loader) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      CommandMenuLoaderWrapper,
      {
        name: loader.name,
        search,
        hook: loader.hook,
        category: loader.category,
        valuePrefix
      },
      loader.name
    ))
  ] });
}
function RecentLoaderRunner({ hook, name, filterNames, onResolved }) {
  (0, import_use_recent_commands.useLoaderCollector)(hook, name, filterNames, onResolved);
  return null;
}
function RecentGroup() {
  const { commands, loaders, recentSet, onResolved } = (0, import_use_recent_commands.useRecentCommands)();
  if (!commands.length && !loaders.length) {
    return null;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_cmdk.Command.Group, { heading: (0, import_i18n.__)("Recent"), children: [
    loaders.map((loader) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      RecentLoaderRunner,
      {
        name: loader.name,
        hook: loader.hook,
        filterNames: recentSet,
        onResolved
      },
      loader.name
    )),
    commands.map((command) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      CommandItem,
      {
        command,
        search: "",
        valuePrefix: "recent-"
      },
      command.name
    ))
  ] });
}
function SuggestionsGroup() {
  const { commands, loaders } = (0, import_data.useSelect)((select) => {
    const { getCommands, getCommandLoaders } = select(import_store.store);
    return {
      commands: getCommands(true),
      loaders: getCommandLoaders(true)
    };
  }, []);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_cmdk.Command.Group, { heading: (0, import_i18n.__)("Suggestions"), children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(CommandList, { search: "", commands, loaders }) });
}
function ResultsGroup({ search }) {
  const { commands, contextualCommands, loaders, contextualLoaders } = (0, import_data.useSelect)((select) => {
    const { getCommands, getCommandLoaders } = select(import_store.store);
    return {
      commands: getCommands(false),
      contextualCommands: getCommands(true),
      loaders: getCommandLoaders(false),
      contextualLoaders: getCommandLoaders(true)
    };
  }, []);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_cmdk.Command.Group, { heading: (0, import_i18n.__)("Results"), children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      CommandList,
      {
        search,
        commands,
        loaders
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      CommandList,
      {
        search,
        commands: contextualCommands,
        loaders: contextualLoaders
      }
    )
  ] });
}
function CommandInput({ search, setSearch }) {
  const commandMenuInput = (0, import_element.useRef)();
  const _value = (0, import_cmdk.useCommandState)((state) => state.value);
  const selectedItemId = _value ? `${ITEM_ID_PREFIX}${_value}` : null;
  (0, import_element.useEffect)(() => {
    commandMenuInput.current.focus();
  }, []);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_cmdk.Command.Input,
    {
      ref: commandMenuInput,
      value: search,
      onValueChange: setSearch,
      placeholder: inputLabel,
      "aria-activedescendant": selectedItemId
    }
  );
}
function CommandMenu() {
  const { registerShortcut } = (0, import_data.useDispatch)(import_keyboard_shortcuts.store);
  const [search, setSearch] = (0, import_element.useState)("");
  const { isOpen: paletteIsOpen, loadersLoading } = (0, import_data.useSelect)(
    (select) => ({
      isOpen: select(import_store.store).isOpen(),
      loadersLoading: (0, import_lock_unlock.unlock)(select(import_store.store)).isLoading()
    }),
    []
  );
  const { open, close } = (0, import_data.useDispatch)(import_store.store);
  (0, import_element.useEffect)(() => {
    registerShortcut({
      name: "core/commands",
      category: "global",
      description: (0, import_i18n.__)("Open the command palette."),
      keyCombination: {
        modifier: "primary",
        character: "k"
      }
    });
  }, [registerShortcut]);
  (0, import_keyboard_shortcuts.useShortcut)(
    "core/commands",
    /** @type {React.KeyboardEventHandler} */
    withIgnoreIMEEvents((event) => {
      if (event.defaultPrevented) {
        return;
      }
      event.preventDefault();
      if (paletteIsOpen) {
        close();
      } else {
        open();
      }
    }),
    {
      bindGlobal: true
    }
  );
  const closeAndReset = () => {
    setSearch("");
    close();
  };
  if (!paletteIsOpen) {
    return false;
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    import_components.Modal,
    {
      className: "commands-command-menu",
      overlayClassName: "commands-command-menu__overlay",
      onRequestClose: closeAndReset,
      __experimentalHideHeader: true,
      size: "medium",
      contentLabel: (0, import_i18n.__)("Command palette"),
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: "commands-command-menu__container", children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_cmdk.Command, { label: inputLabel, loop: true, children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)("div", { className: "commands-command-menu__header", children: [
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            import_icons.Icon,
            {
              className: "commands-command-menu__header-search-icon",
              icon: import_icons.search
            }
          ),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
            CommandInput,
            {
              search,
              setSearch
            }
          )
        ] }),
        /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_cmdk.Command.List, { label: (0, import_i18n.__)("Command suggestions"), children: [
          search && !loadersLoading && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_cmdk.Command.Empty, { children: (0, import_i18n.__)("No results found.") }),
          !search && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(RecentGroup, {}),
          !search && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(SuggestionsGroup, {}),
          search && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(ResultsGroup, { search })
        ] })
      ] }) })
    }
  );
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  CommandMenu,
  isValidIcon
});
