// @wordpress/commands/build/components/use-recent-commands.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/commands/src/components/use-recent-commands.js
var use_recent_commands_exports = {};
__export(use_recent_commands_exports, {
  recordUsage: () => recordUsage,
  useLoaderCollector: () => useLoaderCollector,
  useRecentCommands: () => useRecentCommands
});
module.exports = __toCommonJS(use_recent_commands_exports);
var import_data = require("@wordpress/data");
var import_preferences = require("@wordpress/preferences");
var import_element = require("@wordpress/element");
var import_store = require("../store/index.cjs");
var import_lock_unlock = require("../lock-unlock.cjs");
var MAX_RECENTLY_SAVED = 30;
var MAX_RECENTLY_DISPLAYED = 5;
var EMPTY_ARRAY = [];
var EMPTY_SET = /* @__PURE__ */ new Set();
function recordUsage(name) {
  const current = (0, import_data.select)(import_preferences.store).get(
    "core/commands",
    "recentlyUsed"
  ) ?? [];
  const next = [name, ...current.filter((n) => n !== name)].slice(
    0,
    MAX_RECENTLY_SAVED
  );
  (0, import_data.dispatch)(import_preferences.store).set("core/commands", "recentlyUsed", next);
}
function useLoaderCollector(hook, name, filterNames, onResolved) {
  const { setLoaderLoading } = (0, import_lock_unlock.unlock)((0, import_data.useDispatch)(import_store.store));
  const { isLoading: loading, commands = [] } = hook({ search: "" }) ?? {};
  (0, import_element.useEffect)(() => {
    setLoaderLoading(name, loading);
  }, [setLoaderLoading, name, loading]);
  const filtered = filterNames ? commands.filter((c) => filterNames.has(c.name)) : commands;
  (0, import_element.useEffect)(() => {
    onResolved(name, filtered);
  }, [onResolved, name, filtered]);
  (0, import_element.useEffect)(() => {
    return () => onResolved(name, []);
  }, [onResolved, name]);
}
function useRecentCommands() {
  const {
    contextualCommands,
    staticCommands,
    contextualLoaders,
    staticLoaders,
    recentlyUsedNames = EMPTY_ARRAY
  } = (0, import_data.useSelect)((select) => {
    const { getCommands, getCommandLoaders } = select(import_store.store);
    return {
      contextualCommands: getCommands(true),
      staticCommands: getCommands(false),
      contextualLoaders: getCommandLoaders(true),
      staticLoaders: getCommandLoaders(false),
      recentlyUsedNames: select(import_preferences.store).get(
        "core/commands",
        "recentlyUsed"
      )
    };
  }, []);
  const [resolvedMap, setResolvedMap] = (0, import_element.useState)(() => /* @__PURE__ */ new Map());
  const onResolved = (0, import_element.useCallback)((loaderName, cmds) => {
    setResolvedMap((prev) => {
      const prevCmds = prev.get(loaderName);
      if (prevCmds && prevCmds.length === cmds.length && prevCmds.every((c, i) => c.name === cmds[i].name)) {
        return prev;
      }
      const next = new Map(prev);
      next.set(loaderName, cmds);
      return next;
    });
  }, []);
  const { recentNames, recentSet } = (0, import_element.useMemo)(() => {
    const names = recentlyUsedNames.slice(0, MAX_RECENTLY_DISPLAYED);
    return { recentNames: names, recentSet: new Set(names) };
  }, [recentlyUsedNames]);
  if (!recentlyUsedNames.length) {
    return {
      commands: [],
      loaders: [],
      recentSet: EMPTY_SET,
      onResolved
    };
  }
  const allStaticCommands = [...contextualCommands, ...staticCommands];
  const loaders = [...contextualLoaders, ...staticLoaders];
  const allByName = /* @__PURE__ */ new Map();
  allStaticCommands.forEach((c) => allByName.set(c.name, c));
  for (const cmds of resolvedMap.values()) {
    cmds.forEach((c) => {
      if (!allByName.has(c.name)) {
        allByName.set(c.name, c);
      }
    });
  }
  const commands = recentNames.map((n) => allByName.get(n)).filter(Boolean);
  return { commands, loaders, recentSet, onResolved };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  recordUsage,
  useLoaderCollector,
  useRecentCommands
});
