// @ariakit/react-core/esm/form/form-reset.js
"use client";
import {
  useFormContext
} from "../__chunks/AOHIIJHJ.js";
import {
  useButton
} from "../__chunks/XTCKRCGH.js";
import "../__chunks/EZ4UPVW6.js";
import "../__chunks/3F6D4KUU.js";
import "../__chunks/SWN3JYXT.js";
import {
  useStoreState
} from "../__chunks/SOQQIDO4.js";
import "../__chunks/CTVD4XJH.js";
import {
  createElement,
  createHook,
  forwardRef
} from "../__chunks/L4OUMOCQ.js";
import "../__chunks/W2TDKEPX.js";
import "../__chunks/AZVQSWGA.js";

// src/form/form-reset.tsx
import { invariant } from "@ariakit/core/utils/misc";
var TagName = "button";
var useFormReset = createHook(
  function useFormReset2({ store, ...props }) {
    const context = useFormContext();
    store = store || context;
    invariant(
      store,
      process.env.NODE_ENV !== "production" && "FormReset must be wrapped in a Form component."
    );
    props = {
      type: "reset",
      disabled: useStoreState(store, "submitting"),
      ...props
    };
    props = useButton(props);
    return props;
  }
);
var FormReset = forwardRef(function FormReset2(props) {
  const htmlProps = useFormReset(props);
  return createElement(TagName, htmlProps);
});
export {
  FormReset,
  useFormReset
};
