// @ariakit/react-core/esm/form/form-checkbox.js
"use client";
import {
  useFormControl
} from "../__chunks/ZNASNLV3.js";
import {
  useFormContext
} from "../__chunks/AOHIIJHJ.js";
import "../__chunks/4WQSNMEM.js";
import {
  useCheckboxStore
} from "../__chunks/SNPWGZCA.js";
import {
  useCheckbox
} from "../__chunks/UE4TMJVG.js";
import "../__chunks/EZ4UPVW6.js";
import "../__chunks/3F6D4KUU.js";
import "../__chunks/SWN3JYXT.js";
import "../__chunks/EYKMH5G5.js";
import "../__chunks/ZZQBSCTV.js";
import "../__chunks/SOQQIDO4.js";
import "../__chunks/CTVD4XJH.js";
import {
  createElement,
  createHook,
  forwardRef,
  memo
} from "../__chunks/L4OUMOCQ.js";
import "../__chunks/W2TDKEPX.js";
import "../__chunks/AZVQSWGA.js";

// src/form/form-checkbox.tsx
import { invariant } from "@ariakit/core/utils/misc";
var TagName = "input";
var useFormCheckbox = createHook(
  function useFormCheckbox2({
    store,
    name: nameProp,
    value,
    checked,
    defaultChecked,
    ...props
  }) {
    const context = useFormContext();
    store = store || context;
    invariant(
      store,
      process.env.NODE_ENV !== "production" && "FormCheckbox must be wrapped in a Form component."
    );
    const name = String(nameProp);
    const checkboxStore = useCheckboxStore({
      value: store.useValue(name),
      setValue: (value2) => store == null ? void 0 : store.setValue(name, value2)
    });
    props = useCheckbox({ store: checkboxStore, value, checked, ...props });
    props = useFormControl({
      store,
      name,
      "aria-labelledby": void 0,
      ...props
    });
    return props;
  }
);
var FormCheckbox = memo(
  forwardRef(function FormCheckbox2(props) {
    const htmlProps = useFormCheckbox(props);
    return createElement(TagName, htmlProps);
  })
);
export {
  FormCheckbox,
  useFormCheckbox
};
