// @wordpress/keyboard-shortcuts/build-module/store/selectors.mjs
// packages/keyboard-shortcuts/src/store/selectors.ts
import { createSelector } from "@wordpress/data";
import {
  displayShortcut,
  shortcutAriaLabel,
  rawShortcut
} from "@wordpress/keycodes";
var EMPTY_ARRAY = [];
var FORMATTING_METHODS = {
  /**
   * Display formatting.
   */
  display: displayShortcut,
  /**
   * Raw shortcut formatting.
   */
  raw: rawShortcut,
  /**
   * ARIA label formatting.
   */
  ariaLabel: shortcutAriaLabel
};
function getKeyCombinationRepresentation(shortcut, representation) {
  if (!shortcut) {
    return null;
  }
  return shortcut.modifier ? FORMATTING_METHODS[representation][shortcut.modifier](
    shortcut.character
  ) : shortcut.character;
}
function getShortcutKeyCombination(state, name) {
  return state[name] ? state[name].keyCombination : null;
}
function getShortcutRepresentation(state, name, representation = "display") {
  const shortcut = getShortcutKeyCombination(state, name);
  return getKeyCombinationRepresentation(shortcut, representation);
}
function getShortcutDescription(state, name) {
  return state[name] ? state[name].description : null;
}
function getShortcutAliases(state, name) {
  return state[name] && state[name].aliases ? state[name].aliases : EMPTY_ARRAY;
}
var getAllShortcutKeyCombinations = createSelector(
  (state, name) => {
    return [
      getShortcutKeyCombination(state, name),
      ...getShortcutAliases(state, name)
    ].filter(
      (combination) => !!combination
    );
  },
  (state, name) => [state[name]]
);
var getAllShortcutRawKeyCombinations = createSelector(
  (state, name) => {
    return getAllShortcutKeyCombinations(state, name).map(
      (combination) => getKeyCombinationRepresentation(combination, "raw")
    );
  },
  (state, name) => [state[name]]
);
var getCategoryShortcuts = createSelector(
  (state, categoryName) => {
    return Object.entries(state).filter(([, shortcut]) => shortcut.category === categoryName).map(([name]) => name);
  },
  (state) => [state]
);
export {
  getAllShortcutKeyCombinations,
  getAllShortcutRawKeyCombinations,
  getCategoryShortcuts,
  getShortcutAliases,
  getShortcutDescription,
  getShortcutKeyCombination,
  getShortcutRepresentation
};
