// @wordpress/sync/build-module/providers/http-polling/config.mjs
// packages/sync/src/providers/http-polling/config.ts
import { applyFilters } from "@wordpress/hooks";
var DEFAULT_CLIENT_LIMIT_PER_ROOM = 3;
var ERROR_RETRY_DELAYS_SOLO_MS = [
  2e3,
  4e3,
  8e3,
  12e3
  // Solo: 26s total retry time solo before dialog
];
var ERROR_RETRY_DELAYS_WITH_COLLABORATORS_MS = [
  1e3,
  2e3,
  4e3,
  8e3
  // With collaborators: 15s total retry time before dialog
];
var DISCONNECT_DIALOG_RETRY_MS = 3e4;
var MANUAL_RETRY_INTERVAL_MS = 15e3;
var MAX_ENCODED_UPDATE_SIZE_IN_BYTES = 1 * 1024 * 1024;
var MAX_UPDATE_SIZE_IN_BYTES = Math.floor(MAX_ENCODED_UPDATE_SIZE_IN_BYTES / 4) * 3;
var MAX_ROOMS_PER_REQUEST = 50;
var POLLING_INTERVAL_IN_MS = applyFilters(
  "sync.pollingManager.pollingInterval",
  4e3
  // 4 seconds
);
var POLLING_INTERVAL_WITH_COLLABORATORS_IN_MS = applyFilters(
  "sync.pollingManager.pollingIntervalWithCollaborators",
  1e3
  // 1 second
);
var POLLING_INTERVAL_BACKGROUND_TAB_IN_MS = 25 * 1e3;
export {
  DEFAULT_CLIENT_LIMIT_PER_ROOM,
  DISCONNECT_DIALOG_RETRY_MS,
  ERROR_RETRY_DELAYS_SOLO_MS,
  ERROR_RETRY_DELAYS_WITH_COLLABORATORS_MS,
  MANUAL_RETRY_INTERVAL_MS,
  MAX_ROOMS_PER_REQUEST,
  MAX_UPDATE_SIZE_IN_BYTES,
  POLLING_INTERVAL_BACKGROUND_TAB_IN_MS,
  POLLING_INTERVAL_IN_MS,
  POLLING_INTERVAL_WITH_COLLABORATORS_IN_MS
};
