// @wordpress/sync/build-module/providers/http-polling/polling-manager.mjs
// packages/sync/src/providers/http-polling/polling-manager.ts
import { applyFilters } from "@wordpress/hooks";
import * as Y from "yjs";
import * as encoding from "lib0/encoding";
import * as decoding from "lib0/decoding";
import { removeAwarenessStates } from "y-protocols/awareness";
import * as syncProtocol from "y-protocols/sync";
import {
  DEFAULT_CLIENT_LIMIT_PER_ROOM,
  ERROR_RETRY_DELAYS_SOLO_MS,
  ERROR_RETRY_DELAYS_WITH_COLLABORATORS_MS,
  MAX_ROOMS_PER_REQUEST,
  MAX_UPDATE_SIZE_IN_BYTES,
  POLLING_INTERVAL_IN_MS,
  POLLING_INTERVAL_WITH_COLLABORATORS_IN_MS,
  POLLING_INTERVAL_BACKGROUND_TAB_IN_MS,
  DISCONNECT_DIALOG_RETRY_MS,
  MANUAL_RETRY_INTERVAL_MS
} from "./config.mjs";
import { ConnectionError, ConnectionErrorCode } from "../../errors.mjs";
import {
  SyncUpdateType
} from "./types.mjs";
import {
  base64ToUint8Array,
  createSyncUpdate,
  createUpdateQueue,
  intValueOrDefault,
  postSyncUpdate,
  postSyncUpdateNonBlocking,
  rotateWindow
} from "./utils.mjs";
var POLLING_MANAGER_ORIGIN = "polling-manager";
function isForbiddenError(error) {
  return error?.data?.status === 403;
}
function identifyForbiddenRoom(error, rooms) {
  const message = typeof error.message === "string" ? error.message : "";
  const sortedRooms = [...rooms].sort((a, b) => b.length - a.length);
  for (const room of sortedRooms) {
    if (message.includes(room)) {
      return room;
    }
  }
  return null;
}
function handleForbiddenError(error, requestedRooms) {
  const forbiddenRoom = identifyForbiddenRoom(
    error,
    requestedRooms.map((r) => r.room)
  );
  if (forbiddenRoom) {
    const state = roomStates.get(forbiddenRoom);
    if (state) {
      state.log(
        "Permission denied, unregistering room",
        { error },
        "error",
        true
        // force
      );
      unregisterRoom(forbiddenRoom, { sendDisconnectSignal: false });
    }
    for (const room of requestedRooms) {
      if (room.room === forbiddenRoom || !roomStates.has(room.room)) {
        continue;
      }
      const remainingState = roomStates.get(room.room);
      if (room.updates.length > 0) {
        remainingState.updateQueue.restore(room.updates);
      }
    }
  } else {
    const rooms = [...roomStates.keys()];
    for (const room of rooms) {
      const state = roomStates.get(room);
      if (state) {
        state.log(
          "Permission denied, unregistering room",
          { error },
          "error",
          true
          // force
        );
        unregisterRoom(room, { sendDisconnectSignal: false });
      }
    }
  }
}
var roomStates = /* @__PURE__ */ new Map();
function createDeprecatedCompactionUpdate(updates) {
  const mergeable = updates.filter(
    (u) => [SyncUpdateType.COMPACTION, SyncUpdateType.UPDATE].includes(
      u.type
    )
  ).map((u) => base64ToUint8Array(u.data));
  return createSyncUpdate(
    Y.mergeUpdatesV2(mergeable),
    SyncUpdateType.COMPACTION
  );
}
function createSyncStep1Update(doc) {
  const encoder = encoding.createEncoder();
  syncProtocol.writeSyncStep1(encoder, doc);
  return createSyncUpdate(
    encoding.toUint8Array(encoder),
    SyncUpdateType.SYNC_STEP_1
  );
}
function createSyncStep2Update(doc, step1) {
  const decoder = decoding.createDecoder(step1);
  const encoder = encoding.createEncoder();
  syncProtocol.readSyncMessage(
    decoder,
    encoder,
    doc,
    POLLING_MANAGER_ORIGIN
  );
  return createSyncUpdate(
    encoding.toUint8Array(encoder),
    SyncUpdateType.SYNC_STEP_2
  );
}
function processAwarenessUpdate(state, awareness) {
  const currentStates = awareness.getStates();
  const added = /* @__PURE__ */ new Set();
  const updated = /* @__PURE__ */ new Set();
  const removed = new Set(
    Array.from(currentStates.keys()).filter(
      (clientId) => !state[clientId]
    )
  );
  Object.entries(state).forEach(([clientIdString, awarenessState]) => {
    const clientId = Number(clientIdString);
    if (clientId === awareness.clientID) {
      return;
    }
    if (null === awarenessState) {
      currentStates.delete(clientId);
      removed.add(clientId);
      return;
    }
    if (!currentStates.has(clientId)) {
      currentStates.set(clientId, awarenessState);
      added.add(clientId);
      return;
    }
    const currentState = currentStates.get(clientId);
    if (JSON.stringify(currentState) !== JSON.stringify(awarenessState)) {
      currentStates.set(clientId, awarenessState);
      updated.add(clientId);
    }
  });
  if (added.size + updated.size > 0) {
    awareness.emit("change", [
      {
        added: Array.from(added),
        updated: Array.from(updated),
        // Left blank on purpose, as the removal of clients is handled in the if condition below.
        removed: []
      }
    ]);
  }
  if (removed.size > 0) {
    removeAwarenessStates(
      awareness,
      Array.from(removed),
      POLLING_MANAGER_ORIGIN
    );
  }
}
function processDocUpdate(update, doc, onSync) {
  const data = base64ToUint8Array(update.data);
  switch (update.type) {
    case SyncUpdateType.SYNC_STEP_1: {
      return createSyncStep2Update(doc, data);
    }
    case SyncUpdateType.SYNC_STEP_2: {
      const decoder = decoding.createDecoder(data);
      const encoder = encoding.createEncoder();
      syncProtocol.readSyncMessage(
        decoder,
        encoder,
        doc,
        POLLING_MANAGER_ORIGIN
      );
      onSync();
      return;
    }
    case SyncUpdateType.COMPACTION:
    case SyncUpdateType.UPDATE: {
      Y.applyUpdateV2(doc, data, POLLING_MANAGER_ORIGIN);
    }
  }
}
function checkConnectionLimit(awareness, roomState) {
  if (!roomState.isPrimaryRoom || hasCheckedConnectionLimit) {
    return false;
  }
  hasCheckedConnectionLimit = true;
  const maxClientsPerRoom = applyFilters(
    "sync.pollingProvider.maxClientsPerRoom",
    DEFAULT_CLIENT_LIMIT_PER_ROOM,
    roomState.room
  );
  const clientCount = Object.keys(awareness).length;
  const validatedLimit = intValueOrDefault(
    maxClientsPerRoom,
    DEFAULT_CLIENT_LIMIT_PER_ROOM
  );
  if (clientCount > validatedLimit) {
    roomState.log("Connection limit exceeded", {
      clientCount,
      maxClientsPerRoom: validatedLimit,
      room: roomState.room
    });
    return true;
  }
  return false;
}
var areListenersRegistered = false;
var consecutiveFailures = 0;
var hasCheckedConnectionLimit = false;
var isManualRetry = false;
var hasCollaborators = false;
var isActiveBrowser = "visible" === document.visibilityState;
var isPolling = false;
var isUnloadPending = false;
var pollInterval = POLLING_INTERVAL_IN_MS;
var pollingTimeoutId = null;
var roomOverflowOffset = 0;
function handleBeforeUnload() {
  isUnloadPending = true;
}
function handlePageHide() {
  const rooms = Array.from(roomStates.entries()).map(
    ([room, state]) => ({
      after: 0,
      awareness: null,
      client_id: state.clientId,
      room,
      updates: []
    })
  );
  for (let i = 0; i < rooms.length; i += MAX_ROOMS_PER_REQUEST) {
    postSyncUpdateNonBlocking({
      rooms: rooms.slice(i, i + MAX_ROOMS_PER_REQUEST)
    });
  }
}
function handleVisibilityChange() {
  const wasActive = isActiveBrowser;
  isActiveBrowser = document.visibilityState === "visible";
  if (isActiveBrowser && !wasActive) {
    if (pollingTimeoutId) {
      clearTimeout(pollingTimeoutId);
      pollingTimeoutId = null;
      poll();
    }
  }
}
function selectRoomsForRequest() {
  const allRooms = Array.from(roomStates.values());
  if (allRooms.length <= MAX_ROOMS_PER_REQUEST) {
    return allRooms;
  }
  const primaryRoom = allRooms.find((state) => state.isPrimaryRoom);
  const overflowRooms = allRooms.filter((state) => state !== primaryRoom);
  const overflowSlotsPerRequest = MAX_ROOMS_PER_REQUEST - (primaryRoom ? 1 : 0);
  const { window: overflowSlice, nextOffset } = rotateWindow(
    overflowRooms,
    roomOverflowOffset,
    overflowSlotsPerRequest
  );
  roomOverflowOffset = nextOffset;
  if (primaryRoom) {
    return [primaryRoom, ...overflowSlice];
  }
  return overflowSlice;
}
function poll() {
  isPolling = true;
  pollingTimeoutId = null;
  async function start() {
    if (0 === roomStates.size) {
      isPolling = false;
      return;
    }
    isUnloadPending = false;
    const roomsInRequest = selectRoomsForRequest();
    const payload = {
      rooms: roomsInRequest.map((state) => ({
        after: state.endCursor ?? 0,
        awareness: state.localAwarenessState,
        client_id: state.clientId,
        room: state.room,
        updates: state.updateQueue.get()
      }))
    };
    roomsInRequest.forEach((state) => {
      state.onStatusChange({ status: "connecting" });
    });
    try {
      const { rooms } = await postSyncUpdate(payload);
      consecutiveFailures = 0;
      isManualRetry = false;
      roomsInRequest.forEach((state) => {
        if (roomStates.get(state.room) !== state) {
          return;
        }
        state.onStatusChange({ status: "connected" });
      });
      hasCollaborators = false;
      rooms.forEach((room) => {
        if (!roomStates.has(room.room)) {
          return;
        }
        const roomState = roomStates.get(room.room);
        roomState.endCursor = room.end_cursor;
        if (checkConnectionLimit(room.awareness, roomState)) {
          roomState.onStatusChange({
            status: "disconnected",
            error: new ConnectionError(
              ConnectionErrorCode.CONNECTION_LIMIT_EXCEEDED,
              "Connection limit exceeded"
            )
          });
          unregisterRoom(room.room);
          return;
        }
        roomState.processAwarenessUpdate(room.awareness);
        if (roomState.isPrimaryRoom && Object.keys(room.awareness).length > 1) {
          hasCollaborators = true;
          roomStates.forEach((state) => {
            state.updateQueue.resume();
          });
        }
        const responseUpdates = [];
        for (const update of room.updates) {
          try {
            const response = roomState.processDocUpdate(update);
            if (response) {
              responseUpdates.push(response);
            }
          } catch (error) {
            roomState.log(
              "Failed to apply sync update",
              { error, update },
              "error",
              true
              // force
            );
          }
        }
        roomState.updateQueue.addBulk(responseUpdates);
        if (room.should_compact) {
          roomState.log("Server requested compaction update");
          roomState.updateQueue.clear();
          roomState.updateQueue.add(
            roomState.createCompactionUpdate()
          );
        } else if (room.compaction_request) {
          roomState.log("Server requested (old) compaction update");
          roomState.updateQueue.add(
            createDeprecatedCompactionUpdate(
              room.compaction_request
            )
          );
        }
      });
      if (isActiveBrowser && hasCollaborators) {
        pollInterval = POLLING_INTERVAL_WITH_COLLABORATORS_IN_MS;
      } else if (isActiveBrowser) {
        pollInterval = POLLING_INTERVAL_IN_MS;
      } else {
        pollInterval = POLLING_INTERVAL_BACKGROUND_TAB_IN_MS;
      }
    } catch (error) {
      if (isForbiddenError(error)) {
        handleForbiddenError(error, payload.rooms);
        if (roomStates.size === 0) {
          isPolling = false;
          return;
        }
      } else {
        consecutiveFailures++;
        const retrySchedule = hasCollaborators ? ERROR_RETRY_DELAYS_WITH_COLLABORATORS_MS : ERROR_RETRY_DELAYS_SOLO_MS;
        if (consecutiveFailures <= retrySchedule.length) {
          pollInterval = retrySchedule[consecutiveFailures - 1];
        } else {
          pollInterval = DISCONNECT_DIALOG_RETRY_MS;
        }
        if (isManualRetry) {
          pollInterval = MANUAL_RETRY_INTERVAL_MS;
          isManualRetry = false;
        }
        for (const room of payload.rooms) {
          if (!roomStates.has(room.room)) {
            continue;
          }
          const state = roomStates.get(room.room);
          if (room.updates.length > 0 && state.endCursor > 0) {
            state.updateQueue.clear();
            state.updateQueue.add(state.createCompactionUpdate());
          } else if (room.updates.length > 0) {
            state.updateQueue.restore(room.updates);
          }
          state.log(
            "Error posting sync update, will retry with backoff",
            { error, nextPoll: pollInterval },
            "error",
            true
            // force
          );
        }
        if (!isUnloadPending) {
          const backgroundRetriesFailed = consecutiveFailures > retrySchedule.length;
          roomsInRequest.forEach((state) => {
            if (roomStates.get(state.room) !== state) {
              return;
            }
            state.onStatusChange({
              status: "disconnected",
              canManuallyRetry: true,
              consecutiveFailures,
              backgroundRetriesFailed,
              willAutoRetryInMs: pollInterval
            });
          });
        }
      }
    }
    pollingTimeoutId = setTimeout(poll, pollInterval);
  }
  void start();
}
function registerRoom({
  room,
  doc,
  awareness,
  log,
  onSync,
  onStatusChange
}) {
  if (roomStates.has(room)) {
    return;
  }
  const updateQueue = createUpdateQueue([createSyncStep1Update(doc)]);
  const isPrimaryRoom = 0 === roomStates.size;
  function onAwarenessUpdate() {
    roomState.localAwarenessState = awareness.getLocalState() ?? {};
  }
  function onDocUpdate(update, origin) {
    if (POLLING_MANAGER_ORIGIN === origin) {
      return;
    }
    if (update.byteLength > MAX_UPDATE_SIZE_IN_BYTES) {
      const state = roomStates.get(room);
      if (!state) {
        return;
      }
      state.log("Document size limit exceeded", {
        maxUpdateSizeInBytes: MAX_UPDATE_SIZE_IN_BYTES,
        updateSizeInBytes: update.byteLength
      });
      state.onStatusChange({
        status: "disconnected",
        error: new ConnectionError(
          ConnectionErrorCode.DOCUMENT_SIZE_LIMIT_EXCEEDED,
          "Document size limit exceeded"
        )
      });
      unregisterRoom(room);
      return;
    }
    updateQueue.add(createSyncUpdate(update, SyncUpdateType.UPDATE));
  }
  function unregister() {
    doc.off("updateV2", onDocUpdate);
    awareness.off("change", onAwarenessUpdate);
    updateQueue.clear();
  }
  const roomState = {
    clientId: doc.clientID,
    createCompactionUpdate: () => createSyncUpdate(
      Y.encodeStateAsUpdateV2(doc),
      SyncUpdateType.COMPACTION
    ),
    endCursor: 0,
    isPrimaryRoom,
    localAwarenessState: awareness.getLocalState() ?? {},
    log,
    onStatusChange,
    processAwarenessUpdate: (state) => processAwarenessUpdate(state, awareness),
    processDocUpdate: (update) => processDocUpdate(update, doc, onSync),
    room,
    unregister,
    updateQueue
  };
  doc.on("updateV2", onDocUpdate);
  awareness.on("change", onAwarenessUpdate);
  roomStates.set(room, roomState);
  if (!areListenersRegistered) {
    window.addEventListener("beforeunload", handleBeforeUnload);
    window.addEventListener("pagehide", handlePageHide);
    document.addEventListener("visibilitychange", handleVisibilityChange);
    areListenersRegistered = true;
  }
  if (!isPolling) {
    poll();
  }
}
function unregisterRoom(room, { sendDisconnectSignal = true } = {}) {
  const state = roomStates.get(room);
  if (state) {
    if (sendDisconnectSignal) {
      const rooms = [
        {
          after: 0,
          awareness: null,
          client_id: state.clientId,
          room,
          updates: []
        }
      ];
      postSyncUpdateNonBlocking({ rooms });
    }
    state.unregister();
    roomStates.delete(room);
  }
  if (0 === roomStates.size && areListenersRegistered) {
    window.removeEventListener("beforeunload", handleBeforeUnload);
    window.removeEventListener("pagehide", handlePageHide);
    document.removeEventListener(
      "visibilitychange",
      handleVisibilityChange
    );
    areListenersRegistered = false;
    hasCheckedConnectionLimit = false;
    consecutiveFailures = 0;
    roomOverflowOffset = 0;
  }
}
function retryNow() {
  isManualRetry = true;
  if (pollingTimeoutId) {
    clearTimeout(pollingTimeoutId);
    pollingTimeoutId = null;
    poll();
  }
}
var pollingManager = {
  registerRoom,
  retryNow,
  unregisterRoom
};
export {
  pollingManager
};
