// @wordpress/ui/build/form/primitives/autocomplete/clear.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/ui/src/form/primitives/autocomplete/clear.tsx
var clear_exports = {};
__export(clear_exports, {
  Clear: () => Clear
});
module.exports = __toCommonJS(clear_exports);
var import_autocomplete = require("@base-ui/react/autocomplete");
var import_element = require("@wordpress/element");
var import_icons = require("@wordpress/icons");
var import_i18n = require("@wordpress/i18n");
var import_icon_button = require("../../../icon-button/index.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
var DEFAULT_RENDER = ({
  "aria-label": ariaLabel = (0, import_i18n.__)("Clear"),
  ...props
}) => /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
  import_icon_button.IconButton,
  {
    icon: import_icons.closeSmall,
    size: "small",
    variant: "minimal",
    tone: "neutral",
    label: ariaLabel,
    ...props
  }
);
var Clear = (0, import_element.forwardRef)(
  function Clear2({ render = DEFAULT_RENDER, ...restProps }, ref) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_autocomplete.Autocomplete.Clear,
      {
        ref,
        render,
        ...restProps
      }
    );
  }
);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Clear
});
