// @wordpress/ui/build/form/primitives/autocomplete/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/ui/src/form/primitives/autocomplete/index.ts
var autocomplete_exports = {};
__export(autocomplete_exports, {
  Clear: () => import_clear.Clear,
  Collection: () => import_collection.Collection,
  Empty: () => import_empty.Empty,
  Input: () => import_input.Input,
  InputGroup: () => import_input_group.InputGroup,
  Item: () => import_item.Item,
  List: () => import_list.List,
  ListBody: () => import_list_body.ListBody,
  Popup: () => import_popup.Popup,
  Portal: () => import_portal.Portal,
  Root: () => import_root.Root,
  Value: () => import_value.Value
});
module.exports = __toCommonJS(autocomplete_exports);
var import_clear = require("./clear.cjs");
var import_collection = require("./collection.cjs");
var import_empty = require("./empty.cjs");
var import_input_group = require("./input-group.cjs");
var import_input = require("./input.cjs");
var import_item = require("./item.cjs");
var import_list = require("./list.cjs");
var import_list_body = require("./list-body.cjs");
var import_portal = require("./portal.cjs");
var import_popup = require("./popup.cjs");
var import_root = require("./root.cjs");
var import_value = require("./value.cjs");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Clear,
  Collection,
  Empty,
  Input,
  InputGroup,
  Item,
  List,
  ListBody,
  Popup,
  Portal,
  Root,
  Value
});
