// @ariakit/react-core/cjs/select/select-label.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _CGP2XJO5cjs = require('../__chunks/CGP2XJO5.cjs');
require('../__chunks/7X7JMIHX.cjs');
require('../__chunks/WDV6EJ2U.cjs');
require('../__chunks/VFNUZFXS.cjs');
require('../__chunks/EOGOO37X.cjs');
require('../__chunks/ZYZTL3IQ.cjs');





var _SQLDFLTEcjs = require('../__chunks/SQLDFLTE.cjs');




var _4KGS3DOWcjs = require('../__chunks/4KGS3DOW.cjs');
require('../__chunks/TWX5YNTH.cjs');

// src/select/select-label.tsx
var _misc = require('@ariakit/core/utils/misc');
var TagName = "div";
var useSelectLabel = _SQLDFLTEcjs.createHook.call(void 0, 
  function useSelectLabel2({ store, ...props }) {
    const context = _CGP2XJO5cjs.useSelectProviderContext.call(void 0, );
    store = store || context;
    _misc.invariant.call(void 0, 
      store,
      process.env.NODE_ENV !== "production" && "SelectLabel must receive a `store` prop or be wrapped in a SelectProvider component."
    );
    const id = _4KGS3DOWcjs.useId.call(void 0, props.id);
    const onClickProp = props.onClick;
    const onClick = _4KGS3DOWcjs.useEvent.call(void 0, (event) => {
      onClickProp == null ? void 0 : onClickProp(event);
      if (event.defaultPrevented) return;
      queueMicrotask(() => {
        const select = store == null ? void 0 : store.getState().selectElement;
        select == null ? void 0 : select.focus();
      });
    });
    props = {
      ...props,
      id,
      ref: _4KGS3DOWcjs.useMergeRefs.call(void 0, store.setLabelElement, props.ref),
      onClick,
      style: {
        cursor: "default",
        ...props.style
      }
    };
    return _misc.removeUndefinedValues.call(void 0, props);
  }
);
var SelectLabel = _SQLDFLTEcjs.memo.call(void 0, 
  _SQLDFLTEcjs.forwardRef.call(void 0, function SelectLabel2(props) {
    const htmlProps = useSelectLabel(props);
    return _SQLDFLTEcjs.createElement.call(void 0, TagName, htmlProps);
  })
);



exports.SelectLabel = SelectLabel; exports.useSelectLabel = useSelectLabel;
