// @ariakit/react-core/cjs/select/select-renderer.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _CGP2XJO5cjs = require('../__chunks/CGP2XJO5.cjs');


var _BUZFHEVFcjs = require('../__chunks/BUZFHEVF.cjs');



var _EB6DQJGWcjs = require('../__chunks/EB6DQJGW.cjs');
require('../__chunks/7X7JMIHX.cjs');
require('../__chunks/WDV6EJ2U.cjs');
require('../__chunks/VFNUZFXS.cjs');
require('../__chunks/EOGOO37X.cjs');


var _RDNUVX4Vcjs = require('../__chunks/RDNUVX4V.cjs');
require('../__chunks/ZYZTL3IQ.cjs');



var _SQLDFLTEcjs = require('../__chunks/SQLDFLTE.cjs');
require('../__chunks/4KGS3DOW.cjs');
require('../__chunks/TWX5YNTH.cjs');

// src/select/select-renderer.tsx
var _array = require('@ariakit/core/utils/array');
var _react = require('react');
var TagName = "div";
function getItemObject(item) {
  if (!item || typeof item !== "object") {
    return { value: `${item}` };
  }
  return item;
}
function findIndicesByValue(items, value) {
  var _a;
  const values = _array.toArray.call(void 0, value);
  const indices = [];
  for (const [index, item] of items.entries()) {
    if (indices.length === values.length) break;
    const object = getItemObject(item);
    if (object.value != null && values.includes(object.value)) {
      indices.push(index);
    } else if ((_a = object.items) == null ? void 0 : _a.length) {
      const childIndices = findIndicesByValue(object.items, value);
      if (childIndices.length) {
        indices.push(index);
      }
    }
  }
  return indices;
}
function useSelectRenderer({
  store,
  orientation: orientationProp,
  persistentIndices: persistentIndicesProp,
  items: itemsProp,
  value: valueProp,
  ...props
}) {
  const context = _CGP2XJO5cjs.useSelectContext.call(void 0, );
  store = store || context;
  const items = _RDNUVX4Vcjs.useStoreState.call(void 0, store, (state) => {
    if (!state) return itemsProp;
    if (!state.mounted) return 0;
    return itemsProp != null ? itemsProp : state.items;
  });
  const value = _RDNUVX4Vcjs.useStoreState.call(void 0, store, (state) => valueProp != null ? valueProp : state == null ? void 0 : state.value);
  const valueIndices = _react.useMemo.call(void 0, () => {
    if (!items) return [];
    if (value == null) return [];
    if (typeof items === "number") return [];
    if (!items.length) return [];
    return findIndicesByValue(items, value);
  }, [items, value]);
  const persistentIndices = _react.useMemo.call(void 0, () => {
    if (persistentIndicesProp) {
      return [...persistentIndicesProp, ...valueIndices];
    }
    return valueIndices;
  }, [valueIndices, persistentIndicesProp]);
  return _BUZFHEVFcjs.useCompositeRenderer.call(void 0, {
    store,
    items,
    persistentIndices,
    ...props
  });
}
var SelectRenderer = _SQLDFLTEcjs.forwardRef.call(void 0, function SelectRenderer2(props) {
  const htmlProps = useSelectRenderer(props);
  return _SQLDFLTEcjs.createElement.call(void 0, TagName, htmlProps);
});




exports.SelectRenderer = SelectRenderer; exports.getSelectRendererItem = _EB6DQJGWcjs.getCollectionRendererItem; exports.getSelectRendererItemId = _EB6DQJGWcjs.getCollectionRendererItemId;
