// @wordpress/commands/build-module/components/command-menu.mjs
// packages/commands/src/components/command-menu.js
import { Command, useCommandState } from "cmdk";
import clsx from "clsx";
import { useSelect, useDispatch } from "@wordpress/data";
import {
  useState,
  useEffect,
  useRef,
  isValidElement,
  Component
} from "@wordpress/element";
import { __ } from "@wordpress/i18n";
import {
  Modal,
  TextHighlight,
  __experimentalHStack as HStack,
  privateApis as componentsPrivateApis
} from "@wordpress/components";
import {
  store as keyboardShortcutsStore,
  useShortcut
} from "@wordpress/keyboard-shortcuts";
import { Icon, search as inputIcon, arrowRight } from "@wordpress/icons";
import { store as commandsStore } from "../store/index.mjs";
import { unlock } from "../lock-unlock.mjs";
import {
  recordUsage,
  useLoaderCollector,
  useRecentCommands
} from "./use-recent-commands.mjs";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
var { withIgnoreIMEEvents } = unlock(componentsPrivateApis);
var ITEM_ID_PREFIX = "command-palette-item-";
var inputLabel = __("Search commands and settings");
var CATEGORY_ICONS = {
  view: arrowRight
};
var CATEGORY_LABELS = {
  command: __("Command"),
  view: __("View"),
  edit: __("Edit"),
  action: __("Action"),
  workflow: __("Workflow")
};
function isValidIcon(icon) {
  return !!icon && (typeof icon === "string" || isValidElement(icon) || typeof icon === "function" || icon instanceof Component);
}
function CommandItem({ command, search, category, valuePrefix }) {
  const { close } = useDispatch(commandsStore);
  const commandCategory = category ?? command.category;
  const label = command.searchLabel ?? command.label;
  const value = valuePrefix ? `${valuePrefix}${command.name}` : label;
  return /* @__PURE__ */ jsx(
    Command.Item,
    {
      id: `${ITEM_ID_PREFIX}${value.toLowerCase()}`,
      value,
      keywords: valuePrefix ? [...command.keywords ?? [], label] : command.keywords,
      onSelect: () => {
        recordUsage(command.name);
        command.callback({ close });
      },
      children: /* @__PURE__ */ jsxs(
        HStack,
        {
          alignment: "left",
          className: clsx("commands-command-menu__item", {
            "has-icon": CATEGORY_ICONS[commandCategory] || command.icon
          }),
          children: [
            CATEGORY_ICONS[commandCategory] ? /* @__PURE__ */ jsx(Icon, { icon: CATEGORY_ICONS[commandCategory] }) : isValidIcon(command.icon) && /* @__PURE__ */ jsx(Icon, { icon: command.icon }),
            /* @__PURE__ */ jsx("span", { className: "commands-command-menu__item-label", children: /* @__PURE__ */ jsx(
              TextHighlight,
              {
                text: command.label,
                highlight: search
              }
            ) }),
            CATEGORY_LABELS[commandCategory] && /* @__PURE__ */ jsx("span", { className: "commands-command-menu__item-category", children: CATEGORY_LABELS[commandCategory] })
          ]
        }
      )
    },
    command.name
  );
}
function CommandMenuLoader({ name, search, hook, category, valuePrefix }) {
  const { setLoaderLoading } = unlock(useDispatch(commandsStore));
  const { isLoading: loading, commands = [] } = hook({ search }) ?? {};
  useEffect(() => {
    setLoaderLoading(name, loading);
  }, [setLoaderLoading, name, loading]);
  if (!commands.length) {
    return null;
  }
  return /* @__PURE__ */ jsx(Fragment, { children: commands.map((command) => /* @__PURE__ */ jsx(
    CommandItem,
    {
      command,
      search,
      category: command.category ?? category,
      valuePrefix
    },
    command.name
  )) });
}
function CommandMenuLoaderWrapper({ hook, ...props }) {
  const currentLoaderRef = useRef(hook);
  const [key, setKey] = useState(0);
  useEffect(() => {
    if (currentLoaderRef.current !== hook) {
      currentLoaderRef.current = hook;
      setKey((prevKey) => prevKey + 1);
    }
  }, [hook]);
  return /* @__PURE__ */ jsx(
    CommandMenuLoader,
    {
      hook: currentLoaderRef.current,
      ...props
    },
    key
  );
}
function CommandList({ search, commands, loaders, valuePrefix }) {
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    commands.map((command) => /* @__PURE__ */ jsx(
      CommandItem,
      {
        command,
        search,
        valuePrefix
      },
      command.name
    )),
    loaders.map((loader) => /* @__PURE__ */ jsx(
      CommandMenuLoaderWrapper,
      {
        name: loader.name,
        search,
        hook: loader.hook,
        category: loader.category,
        valuePrefix
      },
      loader.name
    ))
  ] });
}
function RecentLoaderRunner({ hook, name, filterNames, onResolved }) {
  useLoaderCollector(hook, name, filterNames, onResolved);
  return null;
}
function RecentGroup() {
  const { commands, loaders, recentSet, onResolved } = useRecentCommands();
  if (!commands.length && !loaders.length) {
    return null;
  }
  return /* @__PURE__ */ jsxs(Command.Group, { heading: __("Recent"), children: [
    loaders.map((loader) => /* @__PURE__ */ jsx(
      RecentLoaderRunner,
      {
        name: loader.name,
        hook: loader.hook,
        filterNames: recentSet,
        onResolved
      },
      loader.name
    )),
    commands.map((command) => /* @__PURE__ */ jsx(
      CommandItem,
      {
        command,
        search: "",
        valuePrefix: "recent-"
      },
      command.name
    ))
  ] });
}
function SuggestionsGroup() {
  const { commands, loaders } = useSelect((select) => {
    const { getCommands, getCommandLoaders } = select(commandsStore);
    return {
      commands: getCommands(true),
      loaders: getCommandLoaders(true)
    };
  }, []);
  return /* @__PURE__ */ jsx(Command.Group, { heading: __("Suggestions"), children: /* @__PURE__ */ jsx(CommandList, { search: "", commands, loaders }) });
}
function ResultsGroup({ search }) {
  const { commands, contextualCommands, loaders, contextualLoaders } = useSelect((select) => {
    const { getCommands, getCommandLoaders } = select(commandsStore);
    return {
      commands: getCommands(false),
      contextualCommands: getCommands(true),
      loaders: getCommandLoaders(false),
      contextualLoaders: getCommandLoaders(true)
    };
  }, []);
  return /* @__PURE__ */ jsxs(Command.Group, { heading: __("Results"), children: [
    /* @__PURE__ */ jsx(
      CommandList,
      {
        search,
        commands,
        loaders
      }
    ),
    /* @__PURE__ */ jsx(
      CommandList,
      {
        search,
        commands: contextualCommands,
        loaders: contextualLoaders
      }
    )
  ] });
}
function CommandInput({ search, setSearch }) {
  const commandMenuInput = useRef();
  const _value = useCommandState((state) => state.value);
  const selectedItemId = _value ? `${ITEM_ID_PREFIX}${_value}` : null;
  useEffect(() => {
    commandMenuInput.current.focus();
  }, []);
  return /* @__PURE__ */ jsx(
    Command.Input,
    {
      ref: commandMenuInput,
      value: search,
      onValueChange: setSearch,
      placeholder: inputLabel,
      "aria-activedescendant": selectedItemId
    }
  );
}
function CommandMenu() {
  const { registerShortcut } = useDispatch(keyboardShortcutsStore);
  const [search, setSearch] = useState("");
  const { isOpen: paletteIsOpen, loadersLoading } = useSelect(
    (select) => ({
      isOpen: select(commandsStore).isOpen(),
      loadersLoading: unlock(select(commandsStore)).isLoading()
    }),
    []
  );
  const { open, close } = useDispatch(commandsStore);
  useEffect(() => {
    registerShortcut({
      name: "core/commands",
      category: "global",
      description: __("Open the command palette."),
      keyCombination: {
        modifier: "primary",
        character: "k"
      }
    });
  }, [registerShortcut]);
  useShortcut(
    "core/commands",
    /** @type {React.KeyboardEventHandler} */
    withIgnoreIMEEvents((event) => {
      if (event.defaultPrevented) {
        return;
      }
      event.preventDefault();
      if (paletteIsOpen) {
        close();
      } else {
        open();
      }
    }),
    {
      bindGlobal: true
    }
  );
  const closeAndReset = () => {
    setSearch("");
    close();
  };
  if (!paletteIsOpen) {
    return false;
  }
  return /* @__PURE__ */ jsx(
    Modal,
    {
      className: "commands-command-menu",
      overlayClassName: "commands-command-menu__overlay",
      onRequestClose: closeAndReset,
      __experimentalHideHeader: true,
      size: "medium",
      contentLabel: __("Command palette"),
      children: /* @__PURE__ */ jsx("div", { className: "commands-command-menu__container", children: /* @__PURE__ */ jsxs(Command, { label: inputLabel, loop: true, children: [
        /* @__PURE__ */ jsxs("div", { className: "commands-command-menu__header", children: [
          /* @__PURE__ */ jsx(
            Icon,
            {
              className: "commands-command-menu__header-search-icon",
              icon: inputIcon
            }
          ),
          /* @__PURE__ */ jsx(
            CommandInput,
            {
              search,
              setSearch
            }
          )
        ] }),
        /* @__PURE__ */ jsxs(Command.List, { label: __("Command suggestions"), children: [
          search && !loadersLoading && /* @__PURE__ */ jsx(Command.Empty, { children: __("No results found.") }),
          !search && /* @__PURE__ */ jsx(RecentGroup, {}),
          !search && /* @__PURE__ */ jsx(SuggestionsGroup, {}),
          search && /* @__PURE__ */ jsx(ResultsGroup, { search })
        ] })
      ] }) })
    }
  );
}
export {
  CommandMenu,
  isValidIcon
};
