// @wordpress/commands/build-module/components/use-recent-commands.mjs
// packages/commands/src/components/use-recent-commands.js
import {
  useSelect,
  useDispatch,
  select as globalSelect,
  dispatch
} from "@wordpress/data";
import { store as preferencesStore } from "@wordpress/preferences";
import { useCallback, useEffect, useMemo, useState } from "@wordpress/element";
import { store as commandsStore } from "../store/index.mjs";
import { unlock } from "../lock-unlock.mjs";
var MAX_RECENTLY_SAVED = 30;
var MAX_RECENTLY_DISPLAYED = 5;
var EMPTY_ARRAY = [];
var EMPTY_SET = /* @__PURE__ */ new Set();
function recordUsage(name) {
  const current = globalSelect(preferencesStore).get(
    "core/commands",
    "recentlyUsed"
  ) ?? [];
  const next = [name, ...current.filter((n) => n !== name)].slice(
    0,
    MAX_RECENTLY_SAVED
  );
  dispatch(preferencesStore).set("core/commands", "recentlyUsed", next);
}
function useLoaderCollector(hook, name, filterNames, onResolved) {
  const { setLoaderLoading } = unlock(useDispatch(commandsStore));
  const { isLoading: loading, commands = [] } = hook({ search: "" }) ?? {};
  useEffect(() => {
    setLoaderLoading(name, loading);
  }, [setLoaderLoading, name, loading]);
  const filtered = filterNames ? commands.filter((c) => filterNames.has(c.name)) : commands;
  useEffect(() => {
    onResolved(name, filtered);
  }, [onResolved, name, filtered]);
  useEffect(() => {
    return () => onResolved(name, []);
  }, [onResolved, name]);
}
function useRecentCommands() {
  const {
    contextualCommands,
    staticCommands,
    contextualLoaders,
    staticLoaders,
    recentlyUsedNames = EMPTY_ARRAY
  } = useSelect((select) => {
    const { getCommands, getCommandLoaders } = select(commandsStore);
    return {
      contextualCommands: getCommands(true),
      staticCommands: getCommands(false),
      contextualLoaders: getCommandLoaders(true),
      staticLoaders: getCommandLoaders(false),
      recentlyUsedNames: select(preferencesStore).get(
        "core/commands",
        "recentlyUsed"
      )
    };
  }, []);
  const [resolvedMap, setResolvedMap] = useState(() => /* @__PURE__ */ new Map());
  const onResolved = useCallback((loaderName, cmds) => {
    setResolvedMap((prev) => {
      const prevCmds = prev.get(loaderName);
      if (prevCmds && prevCmds.length === cmds.length && prevCmds.every((c, i) => c.name === cmds[i].name)) {
        return prev;
      }
      const next = new Map(prev);
      next.set(loaderName, cmds);
      return next;
    });
  }, []);
  const { recentNames, recentSet } = useMemo(() => {
    const names = recentlyUsedNames.slice(0, MAX_RECENTLY_DISPLAYED);
    return { recentNames: names, recentSet: new Set(names) };
  }, [recentlyUsedNames]);
  if (!recentlyUsedNames.length) {
    return {
      commands: [],
      loaders: [],
      recentSet: EMPTY_SET,
      onResolved
    };
  }
  const allStaticCommands = [...contextualCommands, ...staticCommands];
  const loaders = [...contextualLoaders, ...staticLoaders];
  const allByName = /* @__PURE__ */ new Map();
  allStaticCommands.forEach((c) => allByName.set(c.name, c));
  for (const cmds of resolvedMap.values()) {
    cmds.forEach((c) => {
      if (!allByName.has(c.name)) {
        allByName.set(c.name, c);
      }
    });
  }
  const commands = recentNames.map((n) => allByName.get(n)).filter(Boolean);
  return { commands, loaders, recentSet, onResolved };
}
export {
  recordUsage,
  useLoaderCollector,
  useRecentCommands
};
