// @wordpress/ui/build/utils/use-overlay-scroll-state-attributes.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/ui/src/utils/use-overlay-scroll-state-attributes.ts
var use_overlay_scroll_state_attributes_exports = {};
__export(use_overlay_scroll_state_attributes_exports, {
  SCROLL_CONTAINER_ATTR: () => SCROLL_CONTAINER_ATTR,
  useOverlayScrollStateAttributes: () => useOverlayScrollStateAttributes
});
module.exports = __toCommonJS(use_overlay_scroll_state_attributes_exports);
var import_element = require("@wordpress/element");
var SCROLL_CONTAINER_ATTR = "data-wp-ui-overlay-scroll-container";
var SCROLLED_FROM_TOP_ATTR = "data-wp-ui-overlay-scrolled-from-top";
var SCROLLED_FROM_BOTTOM_ATTR = "data-wp-ui-overlay-scrolled-from-bottom";
var SCROLL_TABBABLE_FLAG_ATTR = "data-wp-ui-overlay-scroll-tabbable";
var SCROLL_END_EPSILON = 1;
function reconcileTabbableFlag(el) {
  if (el.hasAttribute(SCROLL_TABBABLE_FLAG_ATTR) && el.getAttribute("tabindex") !== "0") {
    el.removeAttribute(SCROLL_TABBABLE_FLAG_ATTR);
  }
}
function updateScrollAttributes(el) {
  const { scrollTop, clientHeight, scrollHeight } = el;
  const overflows = scrollHeight - clientHeight > SCROLL_END_EPSILON;
  el.toggleAttribute(SCROLLED_FROM_TOP_ATTR, scrollTop > 0);
  el.toggleAttribute(
    SCROLLED_FROM_BOTTOM_ATTR,
    scrollTop + clientHeight < scrollHeight - SCROLL_END_EPSILON
  );
  reconcileTabbableFlag(el);
  if (overflows) {
    if (!el.hasAttribute(SCROLL_TABBABLE_FLAG_ATTR) && el.getAttribute("tabindex") === null) {
      el.setAttribute("tabindex", "0");
      el.setAttribute(SCROLL_TABBABLE_FLAG_ATTR, "");
    }
  } else if (el.hasAttribute(SCROLL_TABBABLE_FLAG_ATTR)) {
    el.removeAttribute("tabindex");
    el.removeAttribute(SCROLL_TABBABLE_FLAG_ATTR);
  }
}
var HOOK_OWNED_ATTRS = [
  SCROLL_CONTAINER_ATTR,
  SCROLLED_FROM_TOP_ATTR,
  SCROLLED_FROM_BOTTOM_ATTR
];
function cleanupScrollAttributes(el) {
  for (const attr of HOOK_OWNED_ATTRS) {
    el.removeAttribute(attr);
  }
  reconcileTabbableFlag(el);
  if (el.hasAttribute(SCROLL_TABBABLE_FLAG_ATTR)) {
    el.removeAttribute("tabindex");
    el.removeAttribute(SCROLL_TABBABLE_FLAG_ATTR);
  }
}
function useOverlayScrollStateAttributes(onScroll) {
  const [node, setNode] = (0, import_element.useState)(null);
  const ref = (0, import_element.useCallback)((el) => {
    setNode(el);
  }, []);
  (0, import_element.useLayoutEffect)(() => {
    if (!node) {
      return;
    }
    node.setAttribute(SCROLL_CONTAINER_ATTR, "");
    updateScrollAttributes(node);
    if (typeof ResizeObserver === "undefined") {
      return () => {
        cleanupScrollAttributes(node);
      };
    }
    const resizeObserver = new ResizeObserver(() => {
      updateScrollAttributes(node);
    });
    resizeObserver.observe(node);
    for (const child of Array.from(node.children)) {
      resizeObserver.observe(child);
    }
    let mutationObserver;
    if (typeof MutationObserver !== "undefined") {
      mutationObserver = new MutationObserver((records) => {
        for (const record of records) {
          if (record.target === node) {
            for (const added of Array.from(record.addedNodes)) {
              if (added instanceof Element) {
                resizeObserver.observe(added);
              }
            }
            for (const removed of Array.from(
              record.removedNodes
            )) {
              if (removed instanceof Element) {
                resizeObserver.unobserve(removed);
              }
            }
          }
        }
        updateScrollAttributes(node);
      });
      mutationObserver.observe(node, {
        childList: true
      });
    }
    return () => {
      resizeObserver.disconnect();
      mutationObserver?.disconnect();
      cleanupScrollAttributes(node);
    };
  }, [node]);
  const handleScroll = (0, import_element.useCallback)(
    (event) => {
      updateScrollAttributes(event.currentTarget);
      onScroll?.(event);
    },
    [onScroll]
  );
  return { ref, onScroll: handleScroll };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  SCROLL_CONTAINER_ATTR,
  useOverlayScrollStateAttributes
});
